---
title: "idxstats_cbind"
author: "Douglas Yu"
date: "28/01/2018"
output: html_document
---

This file takes the idxstats.txt outputs of the samtools filters and combines into a single large table

This version is the macOS code version. 

```{r}
# rm(list=ls())
library(tidyverse)
# ── Attaching packages ──────────────────────────────────────────────────────────── tidyverse 1.2.1 ──
# ✔ ggplot2 3.0.0.9000     ✔ purrr   0.2.5     
# ✔ tibble  1.4.2          ✔ dplyr   0.7.7     
# ✔ tidyr   0.8.1          ✔ stringr 1.3.1     
# ✔ readr   1.1.1          ✔ forcats 0.3.0     
# ── Conflicts ─────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
# ✖ dplyr::filter() masks stats::filter()
# ✖ dplyr::lag()    masks stats::lag()

library(readxl)
library(lubridate)
library(knitr)
library(beepr)
# library(parallel)
# library(knitr)
# library(lattice)
# library(beeswarm)
# library(sjPlot)
# library(openxlsx)
# library(RColorBrewer)

sessionInfo()
```

```{r notify}
# function to allow macOS system notifications at end of a long command
notify <- function(msgString='Message from R', titleString='Message from R', speakIt=FALSE) {
    cmd <- paste('terminal-notifier -message ', '"', msgString, '"  -title "', titleString, '"', sep='')
    system(cmd)

    if (speakIt) {
        system(paste('say', msgString))
    }
}
```

```{r set paths and filenames for PlatesAB}
setwd("~/Dropbox/Working_docs/Roslin_Greenland/2016/bulk_samples/PlatesAB_EI_20160512/")
getwd()

samtoolsfilter <- "F2308_f0x2"
samtoolsqual <- "q48"

# use this version for minimap2 mapping to 308mitogenomes
# outputs_PlatesAB_F2308_f0x2_q48_minimap2_outputs_20190115_308mitogenomes
# idxstatsgenomecovfolder <- paste0("outputs_PlatesAB_", samtoolsfilter, "_", samtoolsqual, "_minimap2_outputs_20190115_308mitogenomes")  # this is the enclosing folder around the idxstats and genomcov files for a given set of soups

# use this version for COI_barcodes mapping
idxstatsgenomecovfolder <- paste0("outputs_PlatesAB_", samtoolsfilter, "_", samtoolsqual, "_minimap2_outputs_20190116_406barcodes")  # this is the enclosing folder around the idxstats and genomcov files for a given set of soups
# outputs_PlatesAB_F2308_f0x2_q48_minimap2_outputs_20190116_406barcodes

idxstatsfile <- paste0("*", samtoolsfilter, "_", samtoolsqual, "_sorted.bam_idxstats\\.txt")  
    # e.g. "*F2308_f0x2_q1_sorted.bam_idxstats\\.txt"
    # e.g. Sample_IPO3916_A2_F2308_q48_sorted.bam_idxstats.txt
# choose the filename suffix to match the desired samtools filter. precede with * as a wildcard. \\ is used to escape the . before txt (because it is read with grep as "any character" in list.files() below)

genomecovfile <- paste0("*", samtoolsfilter, "_", samtoolsqual, "_sorted_genomecov_d\\.txt\\.gz")  
    # e.g. ""*F2308_f0x2_q1_sorted_genomecov_d\\.txt\\.gz""
# choose the filename pattern to match the desired samtools filter. precede with * as a wildcard for the sample names (e.g. Sample_IPO3916_A10_).      

samplemetadatafolder <- "~/Dropbox/Working_docs/Roslin_Greenland/2017/bulk_samples/ArcDyn_scripts/6_referenceseqs_metadata/PlatesAB_metadata"  # this is the enclosing folder around the sample metadata spreadsheet, which i will place on hpc

samplemetadatafilename <- "Samples_for_sequencing_at_TGAC_Zackenberg_bulk_samples_April_2016_20160512.xlsx"

lysisbufferfilename <- "lysis_buffer_volumes_PlatesAB_A2B2_20180208.xlsx"  # this is based on Yinqiu's lysis buffer datasheet, which i edited a bit to make the sample names consistent so that tables could be joined via "Short_name_of_the_sample".  Also, i replaced missing data or cells with notes with NA
```

Read in all idx_stats files, add metadata to columns, and merge into one big table in long format (vertically by time)
```{r 1. read and tabulate idx files for PlatesAB, eval=FALSE}
# produces a list of unique idx filenames (full filenames but no path)
# use this command if i have a particular folder2 value (e.g. BWA02) to pull from
# idx_files <- list.files(file.path(idxstatsgenomecovfolder, folder2), pattern = "*F2308_f0x2_q1_sorted.bam_idxstats\\.txt")

cat("idxstatsgenomecoverfolder is: ", idxstatsgenomecovfolder)
cat("idxstatsfile search pattern is: ", idxstatsfile)

# idx_files <- list.files(Sys.glob(file.path(idxstatsgenomecovfolder, "BWA??", folder2)), pattern = idxstatsfile, full.names = TRUE) # old version
idx_files <- list.files(Sys.glob(file.path(idxstatsgenomecovfolder)), pattern = idxstatsfile, full.names = TRUE)
head(idx_files); cat("    ", length(idx_files), "total files")
    # Sys.glob is used to generate:  [1] "idxstats_files/BWA01" "idxstats_files/BWA02"
    # list.files is used to look in those folders for files matching "*F2308_f0x2_q1_sorted.bam_idxstats\\.txt"
    # full.names = TRUE is used to return the pathnames
    # \\ is used to escape the . before txt (because it is read with grep as "any character")

# column names of idx files
idx_cols <- c("mitogenome", "mt_length", "mapped_reads", "unmapped_reads")

# function to read_tsv the file and filename and pathname
loadFile1 <- function(x) {
    # read in the four columns and store in df
    df <- read_tsv(x,   # previously file.path(x)
    col_names = idx_cols, na = "NA",
    col_types = cols(
        mitogenome = col_character(),
        mt_length = col_integer(),
        mapped_reads = col_integer(),
        unmapped_reads = col_integer()
        ))

    # read in filename, extract and sub in the first bit of the name, remove any filepaths (using basename) and store in df$sample.  example filename:  
    # Sample_IPO3916_A2_F2308_f0x2_q1_sorted.bam_idxstats.txt # PlatesAB

    # df$sample <- sub("([^\\.]+)\\_sorted.bam_idxstats.txt", "\\1", basename(x))
    df$sample <- sub("Sample_([A-Z,0-9]+_\\D\\d+)_[\\w]+_sorted.bam_idxstats.txt", "\\1", basename(x), perl=TRUE)
    # \\w any letter, digit, underscore
    # \\D any non-digit, \\d any digit

    df$well <- sub("Sample_[A-Z,0-9]+_(\\D\\d+)_[\\w]+_sorted.bam_idxstats.txt", "\\1", basename(x), perl=TRUE)
    # extract well number from filename
    # \\w any letter, digit, underscore

    df$samtools_filter <- sub("Sample_[A-Z,0-9]+_\\D\\d+_(F[\\w]+)_sorted.bam_idxstats.txt", "\\1", basename(x), perl=TRUE)
    # extract samtools parameters from filename

    df$pathname <- x
    # store full pathname in df$pathname

    # output is df
    df
}

# lapply the loadFile() function to all of the idx_files, save the output in idx as a list
idx <- purrr::map(idx_files, loadFile1) # map is equivalent to lapply

# Code from:  Location 6472 in Buffalo, Vince. Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools (Kindle Location 6493). O'Reilly Media. Kindle Edition.

# combine lists into a single dataframe (using do.call(rbind, idx)), reorder columns, delete unmapped_reads column, and remove rows that have "*" in the mitogenome field (the last line of each idxstats table)
idx <- do.call(rbind, idx) %>% dplyr::select(mitogenome, mt_length, mapped_reads, sample, well, samtools_filter, pathname) %>% dplyr::filter(mitogenome != "*")

# str(idx)
# dim(idx)
# sanity checks
idx %>% distinct(well) %>% count()  # number of distinct values of well, as a sanity check:  96
idx %>% distinct(sample) %>% count()  # number of distinct values of sample:  192
idx %>% distinct(pathname) %>% count()  # number of distinct values of pathname:  192
idx %>% distinct(samtools_filter) %>% count() # number of distinct values of samtools_filter: should be 1
idx %>% distinct(samtools_filter)  # show samtools filter value:  should be something like F2308_f0x2_q48

idx_mitogenome_table <- idx %>% group_by(well) %>% distinct(mitogenome) %>% count() %>% arrange(desc(n)); View(idx_mitogenome_table) # should be 311 in all rows (308 mitogenomes + 3 COI_spike)
length(idx$mitogenome) / idx_mitogenome_table[1,2] == idx %>% distinct(sample) %>% count()  # should evaluate to TRUE

```


Read in sample metadata
```{r 2. read sample metadata and create Short_name_of_the_sample}
samplemetadata <- read_excel(file.path(samplemetadatafolder, samplemetadatafilename), sheet = "original", na = "NA")

# use tidyr::separate to parse out metadata from Full_name_of_the_sample
samplemetadata <- samplemetadata %>% tidyr::separate(Full_name_of_the_sample, c("Date", "Plot", "Trap", "Week"), sep = "_", remove = FALSE)
samplemetadata$Date <- as_date(ymd(samplemetadata$Date))
samplemetadata$Year <- year(samplemetadata$Date)
samplemetadata <- samplemetadata %>% dplyr::select(Full_name_of_the_sample, Date, Year, Plot:Extraction_kit_protocol)

# use tidyr::unite to make Short_name_of_the_sample
samplemetadata <- samplemetadata %>% tidyr::unite("Short_name_of_the_sample", c("Year", "Plot", "Trap", "Week"))
```


```{r 2. read lysis buffer info and join to sample metadata}
lysisbufferdata <- read_excel(file.path(samplemetadatafolder, lysisbufferfilename), sheet = "revised_sample_name", na = "NA")
lysisbufferdata <- lysisbufferdata %>% dplyr::select(Short_name_of_the_sample, lysis_buffer_orig_total_ul, lysis_buffer_purified_ul)

samplemetadata <- left_join(samplemetadata, lysisbufferdata) # R message: "Joining, by = "Short_name_of_the_sample""

samplemetadata <- samplemetadata %>% dplyr::mutate(lysis_buffer_proportion=lysis_buffer_orig_total_ul/lysis_buffer_purified_ul)

samplemetadata$lysis_buffer_proportion <- round(samplemetadata$lysis_buffer_proportion, 2) # round to 2 decimal places
```


```{r 3. left_join sample metadata to idx table}
idx_meta <- left_join(idx, samplemetadata) # R message:  "Joining, by = c("sample", "well")"

# use tidyr::separate to parse out metadata from Short_name_of_the_sample
idx_meta <- idx_meta %>% tidyr::separate(Short_name_of_the_sample, c("Year", "Plot", "Trap", "Week"), sep = "_", remove = TRUE)

# sort columns, remove useless columns
idx_meta <- idx_meta %>% dplyr::select(Date, Plot, Trap, Week, mitogenome:lysis_buffer_proportion)

idx_meta <- idx_meta %>% dplyr::select(-row, -column, -sample_reference, -species_name, -barcode_used, -Sample_type, -x.260_230_ratio, -Buffer_used, -Extraction_kit_protocol) %>% dplyr::arrange(Date, Trap)

names(idx_meta)
#  [1] "Date"                       "Plot"                       "Trap"                      
#  [4] "Week"                       "mitogenome"                 "mt_length"                 
#  [7] "mapped_reads"               "sample"                     "well"                      
# [10] "samtools_filter"            "pathname"                   "Full_name_of_the_sample"   
# [13] "Year"                       "Sample_alias"               "ArcDyn_Plate_name"         
# [16] "EI_Plate_name"              "TGAC_barcode"               "Concentration_ng_per_ul"   
# [19] "Sample_volume_ul"           "Quantification_method"      "x.260_280_ratio"           
# [22] "lysis_buffer_orig_total_ul" "lysis_buffer_purified_ul"   "lysis_buffer_proportion"   
```


```{r sanity checks}
idx_meta %>% distinct(Date) %>% count()  # number of distinct values of Date:  70
idx_meta %>% distinct(Plot) %>% count()  # number of distinct values of Plot:  should be 1
idx_meta %>% distinct(Trap) %>% count()  # number of distinct values of Trap:  should be 3 (A,B,C)
idx_meta %>% distinct(Week) %>% count()  # number of distinct values of Week:  18
idx_meta %>% distinct(mitogenome) %>% count()  # number of distinct values of mitogenome:  311 == 308 + 3;  barcodes:  barcodes:  409 == 406 + 3 COI spikes
idx_meta %>% distinct(sample) %>% count()  # number of distinct values of sample:   192 (192 is the number of input idx files)
idx_meta %>% distinct(sample)
idx_meta %>% distinct(Sample_alias) %>% count()  # number of distinct values of Sample_alias:  should be = sample = 192
idx_meta %>% distinct(Sample_alias)
idx_meta %>% distinct(well) %>% count()  # number of distinct values of well:  96
idx_meta %>% distinct(ArcDyn_Plate_name) %>% count()  # number of distinct values of ArcDyn_Plate_name:  should be 2
idx_meta %>% distinct(samtools_filter) %>% count()  # number of distinct values of samtools_filter: should be 1
idx_meta %>% distinct(samtools_filter)   # should only have samtools filter information:  e.g. F2308_f0x2_q1
idx_meta %>% distinct(Full_name_of_the_sample) %>% count()  # number of distinct values of Full_name_of_the_sample: should be 192
idx_meta %>% distinct(Full_name_of_the_sample) %>% arrange(Full_name_of_the_sample) 
```


```{r 4. read in genomecov_d.txt.gz files and add to idx_meta}
# genomecovfile <- "*F2308_f0x2_q1_sorted_genomecov_d\\.txt\\.gz"  # choose the filename pattern to match the desired samtools filter. precede with * as a wildcard for the sample names (e.g. Sample_IPO3916_A10_)

cat("Pattern search is for: ", genomecovfile)

genomecov_files <- list.files(Sys.glob(file.path(idxstatsgenomecovfolder)), pattern = genomecovfile, full.names = TRUE)
head(genomecov_files); cat("    ", length(genomecov_files), "total files")
    # Sys.glob is used to generate:  
        # [1] "minimap2_all_outputs_PlatesAB/Sample_IPO3916_A1_F2308_f0x2_q1_sorted_genomecov_d.txt.gz"
    # list.files is used to look in those folders for files matching "*F2308_f0x2_q1_sorted_genomecov_d\\.txt\\.gz"
    # full.names = TRUE is used to return the pathnames
    # \\ is used to escape the . before txt and gz (because it is read with grep as "any character" in list.files()

# column names of genomecov_files
columnnames <- c("mitogenome", "position", "coverage")

# function to read_tsv the file
loadFile2 <- function(x) {
    df <- read_tsv(gzfile(x),   
        # e.g. "minimap2_all_outputs_PlatesAB/Sample_IPO3916_A1_F2308_f0x2_q1_sorted_genomecov_d.txt.gz"
    col_names = columnnames, na = "NA",
    # col_types = cols(
    #     mitogenome = col_character(),
    #     position = col_integer(),
    #     coverage = col_integer()),
    trim_ws = TRUE
    )

    # read in filename, extract and sub in the first bit of the name, remove any filepaths (using basename) and store in df$sample.  example filename:  Sample_IPO3916_A1_F2308_f0x2_q1_sorted_genomecov_d.txt.gz

    df$sample <- sub("Sample_([A-Z,a-f,0-9]+_\\D\\d+)_[\\w]+_sorted_genomecov_d.txt.gz", "\\1", basename(x), perl=TRUE)
    # \\w any letter, digit, underscore
    # \\D any non-digit, \\d any digit

    df$well <- sub("Sample_[A-Z,a-f,0-9]+_(\\D\\d+)_[\\w]+_sorted_genomecov_d.txt.gz", "\\1", basename(x), perl=TRUE)
    # extract well number from filename
    # \\w any letter, digit, underscore

    df$samtools_filter <- sub("Sample_[\\w]+_(F[\\w]+)_sorted_genomecov_d.txt.gz", "\\1", basename(x), perl=TRUE)
    # extract samtools parameters from filename

    df$pathname_genomecov <- x
    # store full pathname in df$pathname

    df <- df %>% dplyr::group_by(mitogenome) %>% 
        summarise(sum_coverage = sum(coverage), 
                  mean_coverage = mean(coverage), 
                  stddev = sd(coverage), 
                  coefvar = sd(coverage)/mean(coverage), 
                  length = n(), 
                  pct_coverage = sum(coverage>0)/n(), # % of positions that have 1 or more reads mapped 
                  sample = first(sample), 
                  well = first(well), 
                  samtools_filter = first(samtools_filter), 
                  pathname_genomecov = first(pathname_genomecov)
                  )

    # output is df
    df
}

# this step requires ~30 minutes to run on mitogenomes, ~ 3 mins to run on barcodes
# lapply the loadFile2() function to all of the genomecov_files, save the output in genomecovfiles as a list
# map is equivalent to lapply
genomecoverages_summ <- purrr::map(genomecov_files, loadFile2); notify("Your loadfile2 function has finally finished!", "Message from R", speakIt=TRUE)

# rbind the list into a dataframe
genomecoverages_summ <- do.call(rbind, genomecoverages_summ)

# save this file to avoid having to constantly run the map() step
# saveRDS(genomecoverages_summ, "genomecoverages_summ_PlatesAB_20190115_308mitogenomes.RDS", compress = TRUE)
# genomecoverages_summ <- readRDS("genomecoverages_summ_PlatesAB_20190115_308mitogenomes.RDS")

# saveRDS(genomecoverages_summ, "genomecoverages_summ_PlatesAB_20190116_406barcodes.RDS", compress = TRUE)
# genomecoverages_summ <- readRDS("genomecoverages_summ_PlatesAB_20190116_406barcodes.RDS")


# left_join genomecoverages_summ to idx_meta
idx_meta_genomecov <- left_join(idx_meta, genomecoverages_summ) # R message:  Joining, by = c("mitogenome", "sample", "well", "samtools_filter")

idx_meta_genomecov <- idx_meta_genomecov %>% arrange(Date, Trap)

# use grepl() to find mitogenome names that are COI spike names
idx_meta_genomecov$COI_Species <- if_else(grepl("_COI$", idx_meta_genomecov$mitogenome), "COI_Spike", "ArcDyn_Species") # this regex works for both the mitogenome and barcode datasets
# idx_meta_genomecov$COI_Species <- if_else(grepl("ng_COI", idx_meta_genomecov$mitogenome), "COI_Spike", "ArcDyn_Species") # this regex works only for hte mitogenome dataset

# reorder the columns for convenience
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::select(Date:mitogenome, COI_Species, mapped_reads, mt_length:pathname_genomecov) # changed sample to mt_length

names(idx_meta_genomecov)
#  [1] "Date"                       "Plot"                       "Trap"                      
#  [4] "Week"                       "mitogenome"                 "COI_Species"               
#  [7] "mapped_reads"               "mt_length"                  "sample"                    
# [10] "well"                       "samtools_filter"            "pathname"                  
# [13] "Full_name_of_the_sample"    "Year"                       "Sample_alias"              
# [16] "ArcDyn_Plate_name"          "EI_Plate_name"              "TGAC_barcode"              
# [19] "Concentration_ng_per_ul"    "Sample_volume_ul"           "Quantification_method"     
# [22] "x.260_280_ratio"            "lysis_buffer_orig_total_ul" "lysis_buffer_purified_ul"  
# [25] "lysis_buffer_proportion"    "sum_coverage"               "mean_coverage"             
# [28] "stddev"                     "coefvar"                    "length"                    
# [31] "pct_coverage"               "pathname_genomecov"   
```


Move the COI spike data into separate columns in a wide dataset and left_join back to test_idx_meta_genomecov
```{r}
COI_spike_reads <- idx_meta_genomecov %>% dplyr::filter(COI_Species == "COI_Spike") # 192 * 3 = 576
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::filter(COI_Species == "ArcDyn_Species") # make a new genomecov dataset without the COI_Spike rows

COI_spike_reads %>% distinct(Full_name_of_the_sample) %>% count()  # number of distinct values of Full_name_of_the_sample: should be 192
COI_spike_reads %>% distinct(Full_name_of_the_sample) %>% arrange(Full_name_of_the_sample) # number of distinct values of Full_name_of_the_sample: should be 192

# spread out mapped_reads of the spike reads
COI_spike_reads_wide <- COI_spike_reads %>% dplyr::select(sample, Full_name_of_the_sample, mitogenome, mapped_reads) %>%  tidyr::spread(mitogenome, mapped_reads) %>% dplyr::arrange(Full_name_of_the_sample)

names(COI_spike_reads_wide)
# [1] "sample"                                     "Full_name_of_the_sample"                   
# [3] "1-2_Lepidoptera_Bombycidae_Bombyx_mori_COI" "2-1_Coleoptera_Elateridae_COI"             
# [5] "3-1_Coleoptera_Mordellidae_COI"  

# rename columns to remove numbers from in front
names(COI_spike_reads_wide)
# [1] "sample"                                     "Full_name_of_the_sample"                   
# [3] "1-2_Lepidoptera_Bombycidae_Bombyx_mori_COI" "2-1_Coleoptera_Elateridae_COI"             
# [5] "3-1_Coleoptera_Mordellidae_COI"  
# rename columns to remove numbers from in front

# mitogenomes 
COI_spike_reads_wide <- COI_spike_reads_wide %>% dplyr::rename(Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI="1-2_Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI")
COI_spike_reads_wide <- COI_spike_reads_wide %>% dplyr::rename(Coleoptera_Elateridae_40ng_COI="2-1_Coleoptera_Elateridae_40ng_COI")
COI_spike_reads_wide <- COI_spike_reads_wide %>% dplyr::rename(Coleoptera_Mordellidae_20ng_COI="3-1_Coleoptera_Mordellidae_20ng_COI")

# barcodes
COI_spike_reads_wide <- COI_spike_reads_wide %>% dplyr::rename(Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI="1-2_Lepidoptera_Bombycidae_Bombyx_mori_COI")
COI_spike_reads_wide <- COI_spike_reads_wide %>% dplyr::rename(Coleoptera_Elateridae_40ng_COI="2-1_Coleoptera_Elateridae_COI")
COI_spike_reads_wide <- COI_spike_reads_wide %>% dplyr::rename(Coleoptera_Mordellidae_20ng_COI="3-1_Coleoptera_Mordellidae_COI")

names(COI_spike_reads_wide)
# [1] "sample"                                      "Full_name_of_the_sample"                    
# [3] "Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI" "Coleoptera_Elateridae_40ng_COI"             
# [5] "Coleoptera_Mordellidae_20ng_COI"  

# left_join COI_spike_reads_wide to idx_meta_genomecov
idx_meta_genomecov <- left_join(idx_meta_genomecov, COI_spike_reads_wide) # R message:  Joining, by = c("sample", "Full_name_of_the_sample")

names(idx_meta_genomecov)
#  [1] "Date"                                        "Plot"                                       
#  [3] "Trap"                                        "Week"                                       
#  [5] "mitogenome"                                  "COI_Species"                                
#  [7] "mapped_reads"                                "mt_length"                                  
#  [9] "sample"                                      "well"                                       
# [11] "samtools_filter"                             "pathname"                                   
# [13] "Full_name_of_the_sample"                     "Year"                                       
# [15] "Sample_alias"                                "ArcDyn_Plate_name"                          
# [17] "EI_Plate_name"                               "TGAC_barcode"                               
# [19] "Concentration_ng_per_ul"                     "Sample_volume_ul"                           
# [21] "Quantification_method"                       "x.260_280_ratio"                            
# [23] "lysis_buffer_orig_total_ul"                  "lysis_buffer_purified_ul"                   
# [25] "lysis_buffer_proportion"                     "sum_coverage"                               
# [27] "mean_coverage"                               "stddev"                                     
# [29] "coefvar"                                     "length"                                     
# [31] "pathname_genomecov"                          "Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI"
# [33] "Coleoptera_Elateridae_40ng_COI"              "Coleoptera_Mordellidae_20ng_COI"            
```


```{r cleanup files}
rm(genomecoverages_summ)
rm(idx)
rm(idx_meta)
rm(idx_mitogenome_table)
rm(lysisbufferdata)
rm(samplemetadata)
rm(COI_spike_reads)
rm(COI_spike_reads_wide)
# rm(test_idx_meta)
# rm(test_samplemetadata)
# rm(genomecoverages_summ_test)
```


Analyse the COI_spike read numbers and choose which reads to use as a correction factor. 
```{r}
# Calculate the pairwise ratios of the different COI_spike reads
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate(ratio_mordellid20_bombyx10=Coleoptera_Mordellidae_20ng_COI/Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI)
idx_meta_genomecov$ratio_mordellid20_bombyx10 <- round(idx_meta_genomecov$ratio_mordellid20_bombyx10, 1)
ggplot(idx_meta_genomecov, aes(x=ratio_mordellid20_bombyx10)) + geom_density(alpha=0.5)
mean(idx_meta_genomecov$ratio_mordellid20_bombyx10); sd(idx_meta_genomecov$ratio_mordellid20_bombyx10)
# should be 2, but it's 3.69, sd = 0.75

idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate(ratio_elaterid40_bombyx=Coleoptera_Elateridae_40ng_COI/Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI)
idx_meta_genomecov$ratio_elaterid40_bombyx <- round(idx_meta_genomecov$ratio_elaterid40_bombyx, 1)
ggplot(idx_meta_genomecov, aes(x=ratio_elaterid40_bombyx)) + geom_density(alpha=0.5)
mean(idx_meta_genomecov$ratio_elaterid40_bombyx); sd(idx_meta_genomecov$ratio_elaterid40_bombyx)
# should be 4, but it's 7.3, sd = 1.58

idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate(ratio_elaterid40_mordellid20=Coleoptera_Elateridae_40ng_COI/Coleoptera_Mordellidae_20ng_COI)
idx_meta_genomecov$ratio_elaterid40_mordellid20 <- round(idx_meta_genomecov$ratio_elaterid40_mordellid20, 1)
ggplot(idx_meta_genomecov, aes(x=ratio_elaterid40_mordellid20)) + geom_density(alpha=0.5)
mean(idx_meta_genomecov$ratio_elaterid40_mordellid20); sd(idx_meta_genomecov$ratio_elaterid40_mordellid20)
# should be 2, and it's 1.97, sd = 0.07

# Thus, use sum of Coleoptera_Elateridae_40ng_COI and Coleoptera_Mordellidae_20ng_COI to correct reads
# calculate sum of these two COI spike reads for each row
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate(elaterid_mordellid_sum = Coleoptera_Elateridae_40ng_COI + Coleoptera_Mordellidae_20ng_COI)
ggplot(idx_meta_genomecov, aes(x=elaterid_mordellid_sum)) + geom_density(alpha=0.5)
```

Filter out samples that failed to sequence successfully.  I'm only doing this so that the wide table doesn't use a too-small COI_spike correction. 
```{r}
# calculate sum of all COI spike reads for each row
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate(COI_spike_sum = Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI + Coleoptera_Elateridae_40ng_COI + Coleoptera_Mordellidae_20ng_COI)

sum_idx_meta_genomecov <- idx_meta_genomecov %>%
    group_by(Full_name_of_the_sample) %>%
    summarise_at(vars(mapped_reads, COI_spike_sum), funs(sum(as.numeric(.)))) %>%
    arrange(COI_spike_sum)
sum_idx_meta_genomecov <- sum_idx_meta_genomecov %>%
    mutate(species_reads_per_COIspike = mapped_reads/COI_spike_sum)

plot(mapped_reads ~ COI_spike_sum, data = sum_idx_meta_genomecov)

par(mfrow=c(3,2))
hist(sum_idx_meta_genomecov$species_reads_per_COIspike, breaks = 200, xlim = c(0.0000, 0.001))
hist(sum_idx_meta_genomecov$species_reads_per_COIspike, breaks = 200, xlim = c(0.0000, 0.0001))
hist(sum_idx_meta_genomecov$mapped_reads, breaks = 2000, xlim = c(0, 50000))
hist(sum_idx_meta_genomecov$mapped_reads, breaks = 2000, xlim = c(0, 10000))
hist(sum_idx_meta_genomecov$COI_spike_sum, breaks = 1000000, xlim = c(0, 42000000))
hist(sum_idx_meta_genomecov$COI_spike_sum, breaks = 1000000, xlim = c(0, 100000))
par(mfrow=c(1,1))

# filter out samples with COI_spike_sum < 300000 (summed over all 308 mitogenomes or 406 barcodes in the sample).  The largest value of elaterid_mordellid_sum is 7,462,245,896, so this keeps the COI_corr from getting too large (and thus causing the COI_corr reads from getting too small)

sample_names <- sum_idx_meta_genomecov %>% dplyr::filter(COI_spike_sum < 300000) %>% dplyr::select(Full_name_of_the_sample)
samples_to_remove <- sample_names %>% dplyr::pull(Full_name_of_the_sample) # pull out a single variable
samples_to_remove # 10 samples to remove
idx_meta_genomecov <- idx_meta_genomecov %>% filter(!Full_name_of_the_sample %in% samples_to_remove) # 58828 obs mitogenomes, 77546 obs barcodes

# sanity check:  191 samples, which is correct since i removed 1 sample
idx_meta_genomecov %>% distinct(Full_name_of_the_sample) %>% count()  # number of distinct values of Full_name_of_the_sample
```


Correct mapped reads for COI_spike information and lysis buffer proportion.  i'm only doing this for wide-table visualisation
```{r}
# calculate COI_correction factor:  elaterid_mordellid_sum/min(elaterid_mordellid_sum)
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate(COI_corr = elaterid_mordellid_sum/min(elaterid_mordellid_sum))
ggplot(idx_meta_genomecov, aes(x=COI_corr)) + geom_density(alpha=0.5)

# calculate COI_corrected mapped_reads
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate(mapped_reads_COI_corr = mapped_reads / COI_corr)

# calculate lysis_buffer_corrected mapped reads
# add column with the number of mapped reads corrected for the proportion of original lysis buffer used
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate(mapped_reads_COI_lysis_corr = mapped_reads_COI_corr * lysis_buffer_proportion)
idx_meta_genomecov$mapped_reads_COI_lysis_corr <- round(idx_meta_genomecov$mapped_reads_COI_lysis_corr, 0)

# reorder columns and round off mapped_reads_COI_corr
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::select(Date:mapped_reads, mapped_reads_COI_corr, mapped_reads_COI_lysis_corr, mean_coverage, stddev, coefvar, length, mt_length:COI_corr)  
idx_meta_genomecov$mapped_reads_COI_corr <- round(idx_meta_genomecov$mapped_reads_COI_corr, 1)

# generate summed table again
sum_idx_meta_genomecov2 <- idx_meta_genomecov %>% 
    group_by(Full_name_of_the_sample) %>% 
    summarise_at(vars(mapped_reads, mapped_reads_COI_corr, mapped_reads_COI_lysis_corr, sum_coverage, elaterid_mordellid_sum), funs(sum(as.numeric(.)))) %>%
    arrange(mapped_reads_COI_lysis_corr)

names(idx_meta_genomecov)
#  [1] "Date"                                        "Plot"                                       
#  [3] "Trap"                                        "Week"                                       
#  [5] "mitogenome"                                  "COI_Species"                                
#  [7] "mapped_reads"                                "mapped_reads_COI_corr"                      
#  [9] "mapped_reads_COI_lysis_corr"                 "mean_coverage"                              
# [11] "stddev"                                      "coefvar"                                    
# [13] "length"                                      "mt_length"                                  
# [15] "sample"                                      "well"                                       
# [17] "samtools_filter"                             "pathname"                                   
# [19] "Full_name_of_the_sample"                     "Year"                                       
# [21] "Sample_alias"                                "ArcDyn_Plate_name"                          
# [23] "EI_Plate_name"                               "TGAC_barcode"                               
# [25] "Concentration_ng_per_ul"                     "Sample_volume_ul"                           
# [27] "Quantification_method"                       "x.260_280_ratio"                            
# [29] "lysis_buffer_orig_total_ul"                  "lysis_buffer_purified_ul"                   
# [31] "lysis_buffer_proportion"                     "sum_coverage"                               
# [33] "pct_coverage"                                "pathname_genomecov"                         
# [35] "Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI" "Coleoptera_Elateridae_40ng_COI"             
# [37] "Coleoptera_Mordellidae_20ng_COI"             "ratio_mordellid20_bombyx10"                 
# [39] "ratio_elaterid40_bombyx"                     "ratio_elaterid40_mordellid20"               
# [41] "elaterid_mordellid_sum"                      "COI_spike_sum"                              
# [43] "COI_corr"         
```


```{r add DateTrap variable and save idx_meta_genomecov table}
# create variable from Date and Trap(ABC)
# stringr::str_pad(x, width = 2, side = "left", pad = "0")  # to ensure that the date always has two digits, e.g. 07.
idx_meta_genomecov$DateTrap <- str_c("Date", year(idx_meta_genomecov$Date), str_pad(month(idx_meta_genomecov$Date), 2, side = "left", pad = "0"), str_pad(day(idx_meta_genomecov$Date), 2, side = "left", pad = "0"), idx_meta_genomecov$Trap, sep = "_")

# in coefvar column, replace "NaN" string to "NA" and then coerce column back to numeric
idx_meta_genomecov$coefvar <- idx_meta_genomecov$coefvar %>%  str_replace("NaN", "NA") %>% as.numeric()

# these tables check that i only have one row for each Full_name_of_the_sample, DateTrap, and pathname
# ideally, there should be a 1 in all rows of column "n"
idxstats_by_Full_name_of_the_sample <- idx_meta_genomecov %>% dplyr::group_by(Full_name_of_the_sample) %>% dplyr::distinct(pathname) %>% dplyr::count() %>% dplyr::arrange(desc(n)); View(idxstats_by_Full_name_of_the_sample)

# ideally, there should be a 1 in all rows of column "n"
idxstats_by_DateTrap <- idx_meta_genomecov %>% dplyr::group_by(DateTrap) %>% dplyr::distinct(pathname) %>% dplyr::count() %>% dplyr::arrange(desc(n));  View(idxstats_by_DateTrap)

# ideally, there should be a 1 in all rows of column "n"
idxstats_by_pathname <- idx_meta_genomecov %>% dplyr::group_by(pathname) %>% dplyr::distinct(pathname) %>% dplyr::count() %>% dplyr::arrange(desc(n));  View(idxstats_by_pathname)
# i originally found that TrapC_2013_Jul_22 occurred once for week 30 and once for week 29.  I checked the original Excel worksheet and found that 2013JUL22_Art3_TrapC_Wk29 should be 2013JUL15_Art3_TrapC_Wk29.  Thus, the TrapC_2013_Jul_22/week29 combination shouldn't exist.  I have fixed the Excel file.


# write_tsv(idx_meta_genomecov, "idx_meta_genomecov_PlatesAB_20190115_308mitogenomes.txt") # so if i need to read in a file during debuggging, here it is. This is the file for HMSC analysis.
# write_tsv(idx_meta_genomecov, "idx_meta_genomecov_PlatesAB_20190116_406barcodes.txt") # so if i need to read in a file during debuggging, here it is. This is the file for HMSC analysis.

# used for debugging, in case i screw up the dataset while writing code
# idx_meta_genomecov <- read_delim("idx_meta_genomecov_PlatesAB_20190115_308mitogenomes.txt", delim = "\t", escape_double = FALSE, trim_ws = TRUE)
# idx_meta_genomecov <- read_delim("idx_meta_genomecov_PlatesAB_20190116_406barcodes.txt", delim = "\t", escape_double = FALSE, trim_ws = TRUE)
# if i need to read from a folder
# # idx_meta_genomecov <- read_delim("AB_outputs_20180214_F2308_f0x2_q60/idx_meta_genomecov_PlatesAB_20190115_308mitogenomes.txt", delim = "\t", escape_double = FALSE, trim_ws = TRUE)
```

Make wide table for visualisation
```{r}
# Make wide table for visualisation (left to right should follow temporal order (DateTrap))
# mapped_reads_COI_lysis_corr
idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide <- idx_meta_genomecov %>% dplyr::select(DateTrap, mitogenome, mapped_reads_COI_lysis_corr) %>% dplyr::arrange(DateTrap) %>% tidyr::spread(DateTrap, mapped_reads_COI_lysis_corr) %>% dplyr::arrange(mitogenome)

# uncomment to save table
# write_tsv(idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide, "idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide_PlatesAB_20190115_308mitogenomes.txt")
# write_tsv(idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide, "idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide_PlatesAB_20190116_406barcodes.txt")

# make wide table with pct_coverage > pctcovmin > 0.1
plot(idx_meta_genomecov$pct_coverage, idx_meta_genomecov$coefvar)
# set a minimum value of coefvar, above which read numbers are set to -
hist(idx_meta_genomecov$pct_coverage, breaks = 2000, xlim = c(0.000, 1), ylim = c(0, 400)) # reasonable cutoff at 2.0? 3.0?
pctcovmin <- 0.10
idx_meta_genomecov$mapped_reads_COI_lysis_corr_pctcovmin <- ifelse(idx_meta_genomecov$pct_coverage >= pctcovmin, idx_meta_genomecov$mapped_reads_COI_lysis_corr, 0)
# change column order
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::select(Date:mapped_reads_COI_lysis_corr, mapped_reads_COI_lysis_corr_pctcovmin, mean_coverage:DateTrap)

# not necessary
# # replace NA values with 0s
# idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate_at(vars(mapped_reads_COI_lysis_corr_pctcovmin), funs(replace(., is.na(.), 0))) # the . stands in for the variables that are being assessed

# hist(idx_meta_genomecov$coefvar, breaks = 2000, xlim = c(0.0000, 5)) # reasonable cutoff at 2.0? 3.0?
# # set a minimum value of coefvar, above which read numbers are set to -
# coefvarmin <- 3.0
# idx_meta_genomecov$mapped_reads_COI_lysis_corr_coefvarmin <- ifelse(idx_meta_genomecov$coefvar <= coefvarmin, idx_meta_genomecov$mapped_reads_COI_lysis_corr, 0)
# # change column order
# idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::select(Date:mapped_reads_COI_lysis_corr, mapped_reads_COI_lysis_corr_coefvarmin, mean_coverage:DateTrap)
# # replace NA values with 0s
# idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate_at(vars(mapped_reads_COI_lysis_corr_coefvarmin), funs(replace(., is.na(.), 0))) # the . stands in for the variables that are being assessed

# mapped_reads_COI_lysis_corr_pctcovmin # set read number = 0 if coefvar > pctcovmin (e.g. 1.8)
idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide <- idx_meta_genomecov %>% dplyr::select(DateTrap, mitogenome, mapped_reads_COI_lysis_corr_pctcovmin) %>% dplyr::arrange(DateTrap) %>% tidyr::spread(DateTrap, mapped_reads_COI_lysis_corr_pctcovmin) %>% dplyr::arrange(mitogenome)

# uncomment to save table
# write_tsv(idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide, "idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide_PlatesAB_20190115_308mitogenomes.txt")
# write_tsv(idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide, "idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide_PlatesAB_20190116_406barcodes.txt")

```

```{r, eval=FALSE}
# purl("idxstats_tabulate_macOS.Rmd")
```


## Next steps:

1. DONE Read in all idx_stats files, add sample metadata to columns, and merge into one table in long format (vertically by time)
2. DONE merge with sample excel workbook to add the sample date information
3. DONE go through the genomecov files and calculate the mean and standard deviation of coverage for each mitogenome/sample and add the columns to the idx table
4. DONE make wide table using tidyr (the column data format should be DateTrap (e.g. 1998_08_05_TrapA) in temporal order).
5. DONE confirm that i have the correct number of samples (compare with number of input files) e.g. PlatesAB only have 176 bam files (answer:  i only initially downloaded 176 bam files for code testing)
6. DONE Debug the bug that causes sample to have repeated lines, using a tabulate command to check that I don't have repeated lines.  Answer: re-run the minimap2/samtools/bedtools scripts and redownload the idxstats files.  i put the same idxstats files into multiple BWA folders in my test folders
7. DONE platesGH:  combine_fastq, fastQC, multiqc (using PKG-ENQ-2379-Data_Transfer-PSEQ-1586-trimmed), remove orig fastq.gz files.  
8. DONE platesA2B2: combine_fastq, fastQC, multiqc, remove orig fastq.gz files
9. DONE minimap2/samtools for platesGH
10. DONE minimap2/samtools for platesA2B2
11. DONE create lysis buffer datasets, download idxstats and genomecov data, run R code on Plates AB, A2B2, EF, GH

12. Compare PlatesAB and A2B2 WITH YINQIU'S BWA_results_20161025.xlsx.  
13. Compare PlatesGH, PlatesEF with the positive controls, the seasonal patterns, the COI spikes
14. remove original, noncombined fastq.gz files from A2B2 and GH when i'm done mapping and checking.

15. make textfile for taxonomy (class, order, family, genus, species) of each mitogenome and join to idx_meta_genomecov

16. visualise minimap2 bam files and check that i am not mapping to the Ns between the protein-coding genes. SHOULD I SUBTRACT A FIXED NUMBER FROM EACH MITOGENOME LENGTH TO CALCULATE THE CORRECT LENGTH OF THE CODING PORTION?  




names(idx_meta_genomecov)
 [1] "Date"                                        "Plot"                                       
 [3] "Trap"                                        "Week"                                       
 [5] "mitogenome"                                  "COI_Species"                                
 [7] "mapped_reads"                                "mapped_reads_COI_corr"                      
 [9] "mapped_reads_COI_lysis_corr"                 "mean_coverage"                              
[11] "stddev"                                      "coefvar"                                    
[13] "length"                                      "mt_length"                                  
[15] "sample"                                      "well"                                       
[17] "samtools_filter"                             "pathname"                                   
[19] "Full_name_of_the_sample"                     "Year"                                       
[21] "Sample_alias"                                "ArcDyn_Plate_name"                          
[23] "EI_Plate_name"                               "TGAC_barcode"                               
[25] "Concentration_ng_per_ul"                     "Sample_volume_ul"                           
[27] "Quantification_method"                       "x.260_280_ratio"                            
[29] "lysis_buffer_orig_total_ul"                  "lysis_buffer_purified_ul"                   
[31] "lysis_buffer_proportion"                     "sum_coverage"                               
[33] "pathname_genomecov"                          "Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI"
[35] "Coleoptera_Elateridae_40ng_COI"              "Coleoptera_Mordellidae_20ng_COI"            
[37] "ratio_mordellid20_bombyx10"                  "ratio_elaterid40_bombyx"                    
[39] "ratio_elaterid40_mordellid20"                "elaterid_mordellid_sum"                     
[41] "COI_corr"   

### Metadata:
From idxstats.txt
Date:  Date of ArcDyn Sample
Plot:  Zackenberg Plot (only Art3)
Trap:  Window trap within Art3 (A, B, C)
Week:  Week of the year (alternative to Date)
mitogenome:  name of mitogenome sequencer (from reference fasta file used by minimap2)
mt_length:  length of the mitogenome (including Ns between the genes) used by minimap2
mapped_reads:  number of reads mapped to each mitogenome by minimap2 and filtered through samtools
sample:  unique sample name derived from each sample's foldername (e.g. Sample_IPO3916_A1 -> IPO3916_A1). IPO3916 refers to the Plate and A1 refers to the well. I use this to link to my ArcDyn sample metadata.  In other sequencing runs, sample takes the format:  PlateH_F6 (Plate_Well).
EI_sample_name:  sample name based on barcode:  e.g. PrO1747_S142_DNA.  I don't need to use these
well:  well of the Plate
samtools_filter:  parameter used by samtools view to filter out badly mapped reads (e.g. -F2308 -f0x2 -q1 -> F2308_f0x2_q1)
pathname:  path to the idx_stats file

### From Excel spreadsheet that ArcDyn uses to record metadata for each sample
Full_name_of_the_sample:  full name of the sample from the ArcDyn metadata (e.g. 1999JUL08_Art3_TrapA_Wk27)
Short_name_of_the_sample:  different version using only year (e.g. 1997_Art3_TrapA_Wk27)
Sample_alias:  code for the Year_Trap_Week (e.g. 98_B_33)
ArcDyn_Plate_name:  ArcDyn plate name (e.g. Anteater, which is Plate A)
EI_Plate_name:  name of plate given by Earlham Institute (e.g. IPO3916)
TGAC_Barcode:  number on the barcode sticker that i placed on the plate (SAM27529_PRO 1323_S1_gDNA).  this links back to the photo i take of each plate after it is loaded.
Concentration_ng_per_ul:  DNA concentration as it is given to EI
Sample_volume_ul:  how much volume of liquid given to EI per sample in each well
Quantification method: the method ArcDyn used to quantify DNA amount
x.260_280_ratio:  the 260/280 ratio ArcDyn measured for the sample's DNA

### lysis buffer information
lysis_buffer_orig_total_ul:  The total amount of lysis buffer used to digest the sample.  More lysis buffer is used for samples with higher biomass
lysis_buffer_purified_ul:  The amount of lysis buffer used to extract DNA. 

### From bedtools/genomecov -d output file
sum_coverage:  sum of the coverage per nucleotide = the total bases of all the reads that mapped to each mitogenome
mean_coverage:  sum/length of the mitogenome that was mapped to
stddev:  standard deviation of the coverage (e.g. is the mapping even or uneven?)
coefvar:  stddev/mean_coverage.  another measure of unevenness
length:  length of the mitogenome being mapped to.  should be the same as mt_length
pathname_genomecov:  path to the bedtools genomecov -d output file

### Calculated columns
Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI:  reads that mapped to this COI spike sequence
Coleoptera_Elateridae_40ng_COI:  reads that mapped to this COI spike sequence
Coleoptera_Mordellidae_20ng_COI:  reads that mapped to this COI spike sequence
COI_spike_sum:  sum of Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI, Coleoptera_Elateridae_40ng_COI, Coleoptera_Mordellidae_20ng_COI
ratio_mordellid20_bombyx10:  ratio of mordellid to bombyx reads, should be 2
ratio_elaterid40_bombyx:  ratio of elaterid to bombyx reads, should be 4
ratio_elaterid40_mordellid20:  ratio of elaterid to mordellid reads, should be 2
elaterid_mordellid_sum:  sum of elaterid and mordellid reads, used for COI_spike correction
COI_corr:  COI_spike correction factor:  elaterid_mordellid_sum/min(elaterid_mordellid_sum)
lysis_buffer_proportion:  lysis buffer correction factor:  lysis_buffer_purified_ul/lysis_buffer_orig_total_ul
mapped_reads_COI_corr:  mapped reads after correction with COI_corr
mapped_reads_COI_lysis_corr:  mapped reads after correction with COI_corr and lysis_buffer
COI_Species: "COI_Spike", "ArcDyn_Species".   ArcDyn_Species are the ArcDyn soups, and COI_Spike are the spiked in species.  Used to separate out the spikes to correct mapped reads
mock_ArcDyn:  "arcdyn", "control", "mock".  arcdyn is ArcDyn soups.  Control is COI negative controls. mock is mock soups (positive controls)
DateTrap:  format:  Date_1998_08_26_TrapA.  Used to sort the columns in the wide format
YearWeekTrap:  format: 1997_Wk27_TrapA.  Used to sort the columns in the wide format when the Month and Date information not available


Don't use this code.  I wrote this when i tried to do the COI spike corrections with a wide table.  Wrong idea.   
```{r COI_spike corrections on wide table, eval=FALSE, include=FALSE}

# used for debugging, in case i screw up the dataset while writing code
# idx_meta_genomcov_mapped_reads_lysis_corr_wide <- read_delim("idx_meta_genomcov_mapped_reads_lysis_corr_wide.txt", delim = "\t", escape_double = FALSE, trim_ws = TRUE)

# pull out the COI spike rows
bombyx_corr <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% dplyr::filter(str_detect(.$mitogenome, '1-2_Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI'))

mordellid_corr <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% dplyr::filter(str_detect(.$mitogenome, '3-1_Coleoptera_Mordellidae_20ng_COI'))

elaterid_corr <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% dplyr::filter(str_detect(.$mitogenome, '2-1_Coleoptera_Elateridae_40ng_COI'))

# calculate the ratios of the COI spikes and bind the rows to the original dataset
bombyx_bombyx <- as.data.frame(as.numeric(t(bombyx_corr[,-1]))/as.numeric(t(bombyx_corr[,-1])))
colnames(bombyx_bombyx)[1] <- "bombyx_bombyx"
bombyx_bombyx <- as.data.frame(t(bombyx_bombyx)) %>% rownames_to_column(var="mitogenome")

mordellid_bombyx <- data_frame(as.numeric(t(mordellid_corr[,-1]))/as.numeric(t(bombyx_corr[,-1])))
colnames(mordellid_bombyx)[1] <- "mordellid20_bombyx10"
mordellid_bombyx <- as.data.frame(t(mordellid_bombyx)) %>% rownames_to_column(var="mitogenome")

elaterid_bombyx <- data_frame(as.numeric(t(elaterid_corr[,-1]))/as.numeric(t(bombyx_corr[,-1])))
colnames(elaterid_bombyx)[1] <- "elaterid40_bombyx"
elaterid_bombyx <- as.data.frame(t(elaterid_bombyx)) %>% rownames_to_column(var="mitogenome")

elaterid_mordellid <- data_frame(as.numeric(t(elaterid_corr[,-1]))/as.numeric(t(mordellid_corr[,-1])))
colnames(elaterid_mordellid)[1] <- "elaterid40_mordellid20"
elaterid_mordellid <- as.data.frame(t(elaterid_mordellid)) %>% rownames_to_column(var="mitogenome")

# bind together and then bind onto bottom of genomecov wide dataset
COI_spike <- bind_rows(bombyx_bombyx, mordellid_bombyx, elaterid_bombyx, elaterid_mordellid)
colnames(COI_spike) <- colnames(idx_meta_genomcov_mapped_reads_lysis_corr_wide)
idx_meta_genomcov_mapped_reads_lysis_corr_wide <- bind_rows(idx_meta_genomcov_mapped_reads_lysis_corr_wide, COI_spike)

# check pairwise ratios of COI spikes
# should be 2, but it's 3.7
ratio_mordellid20_bombyx10 <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% filter(str_detect(.$mitogenome, 'mordellid20_bombyx10')) %>% dplyr::select(-mitogenome) %>% t() %>% as.numeric() %>% mean(na.rm = TRUE) %>% round(1); ratio_mordellid20_bombyx10

# should be 4, but it's 7.3
ratio_elaterid40_bombyx <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% filter(str_detect(.$mitogenome, 'elaterid40_bombyx')) %>% dplyr::select(-mitogenome) %>% t() %>% as.numeric() %>% mean(na.rm = TRUE) %>% round(1); ratio_elaterid40_bombyx

# should be 2, and it's 1.9
ratio_elaterid40_mordellid20 <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% filter(str_detect(.$mitogenome, 'elaterid40_mordellid20')) %>% dplyr::select(-mitogenome) %>% t() %>% as.numeric() %>% mean(na.rm = TRUE) %>% round(1); ratio_elaterid40_mordellid20

# Use the sum of elaterid and mordellid reads to do the correction.  Bombyx didn't show up

elaterid_mordellid_sum <- as.data.frame(as.numeric(t(elaterid_corr[,-1])) + as.numeric(t(mordellid_corr[,-1])))
colnames(elaterid_mordellid_sum)[1] <- "elaterid_mordellid_sum"
elaterid_mordellid_sum <- as.data.frame(t(elaterid_mordellid_sum)) %>% rownames_to_column(var="mitogenome")
colnames(elaterid_mordellid_sum) <- colnames(idx_meta_genomcov_mapped_reads_lysis_corr_wide)
idx_meta_genomcov_mapped_reads_lysis_corr_wide <- bind_rows(idx_meta_genomcov_mapped_reads_lysis_corr_wide, elaterid_mordellid_sum)

elaterid_mordellid_min <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% filter(str_detect(.$mitogenome, 'elaterid_mordellid_sum')) %>% dplyr::select(-mitogenome)  %>% t() %>% as.numeric() %>% min(na.rm = TRUE)
elaterid_mordellid_min # 115100.3 mapped_reads

# now calculate the column-wise correction factor COI_corr:  elaterid_mordellid_sum/elaterid_mordellid_min
COI_corr <- as.data.frame(as.numeric(t(elaterid_mordellid_sum[,-1]))/elaterid_mordellid_min)
colnames(COI_corr)[1] <- "COI_corr"
COI_corr <- as.data.frame(t(COI_corr)) %>% rownames_to_column(var="mitogenome")
colnames(COI_corr) <- colnames(idx_meta_genomcov_mapped_reads_lysis_corr_wide)
idx_meta_genomcov_mapped_reads_lysis_corr_wide <- bind_rows(idx_meta_genomcov_mapped_reads_lysis_corr_wide, COI_corr)


# pull out COI_corr and divide it into mapped_reads (using a long format dataset) and generate the new wide dataset with lysis and COI corrected reads
COI_corr_t <- t(COI_corr) %>% as.numeric()
%>% as.data.frame()
test <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% gather(date, mapped_reads_lysis_corr, -mitogenome)


```
