---
title: "datafiles_for_ArcDyn_methods_paper"
author: "Douglas Yu"
date: "17/01/2019"
output: html_document
---

Make 10 files for publication. On 20 Dec 2018, Otso Ovaskainen requested that i create 7 datafiles out of the one datafile for publication in the methods paper (and include the total number of reads per sample fastq file)

```{r packages}
library(tidyverse) 
library(readxl) # read excel spreadsheets
library(hablar) # data type conversion
library(arsenal) # utilities
```

```{r reformat seqkit stats output}
# A2B2 resequenced (most of) the same samples as AB, but fortunately, Earlham Institute coded the AB samples as IPO3916/7_A1 and the A2B2 samples as PlateA/B_A1. So I can use this column to join with the other metadata 
pathtofastqseqdata = "~/Dropbox/Working_docs/Roslin_Greenland/" 

# PlatesAB
# sample identifier in soupdata has form IPO3916_A1, where IPO3916 is PlateA, IPO3917 is PlateB
fastq_read_counts_PlatesAB <- read_table(file.path(pathtofastqseqdata, "2016/bulk_samples/PlatesAB_EI_20160512/fastq_read_counts_PlatesAB.txt"))

fastq_read_counts_PlatesAB <- fastq_read_counts_PlatesAB %>%
    tidyr::separate(file, into = c(NA, "gpfs", "home", "b042", "greenland", "plates", "platessub", "BWA", "sample", "fastqfile"), sep = "/", remove = TRUE) %>%
    mutate_at("sample", str_remove, "Sample_") %>% # remove "Sample_" from beginning of string
    dplyr::select(sample, tot_num_seqs = num_seqs) %>%
    dplyr::group_by(sample) %>%
    summarise(tot_num_seqs = sum(tot_num_seqs))

# PlatesA2B2
# sample identifier in soupdata has form PlateA_A1
fastq_read_counts_PlatesA2B2 <- read_table(file.path(pathtofastqseqdata, "2017/bulk_samples/PlatesA2B2/fastq_read_counts_PlatesA2B2.txt"))

fastq_read_counts_PlatesA2B2 <- fastq_read_counts_PlatesA2B2 %>%
    tidyr::separate(file, into = c(NA, "gpfs", "home", "b042", "greenland", "plates", "platessub", "BWA", "sample", "fastqfile"), sep = "/", remove = TRUE) %>%
    mutate_at("sample", str_remove, "Sample_PRO1322_") %>% # remove "Sample_PRO1322_" from beginning of string
    dplyr::select(sample, tot_num_seqs = num_seqs) %>%
    dplyr::group_by(sample) %>%
    summarise(tot_num_seqs = sum(tot_num_seqs))

# merge the files into one file EF
fastq_read_counts_PlatesEF <- read_table(file.path(pathtofastqseqdata, "2017/bulk_samples/PlatesEF/fastq_read_counts_PlatesEF.txt"))

fastq_read_counts_PlatesEF <- fastq_read_counts_PlatesEF %>%
    tidyr::separate(file, into = c(NA, "gpfs", "home", "b042", "greenland", "plates", "platessub", "BWA", "sample", "fastqfile"), sep = "/", remove = TRUE) %>%
    mutate_at("sample", str_remove, "_[ACGT]{9}-[ACGT]{6}") %>% # remove index sequence from end of string, e.g. _TAGGTTAGG-GCCAAT
    dplyr::select(sample, tot_num_seqs = num_seqs) %>%
    dplyr::group_by(sample) %>%
    summarise(tot_num_seqs = sum(tot_num_seqs))

# merge the files into one file GH
fastq_read_counts_PlatesGH <- read_table(file.path(pathtofastqseqdata, "2017/bulk_samples/PlatesGH/fastq_read_counts_PlatesGH.txt"))

fastq_read_counts_PlatesGH <- fastq_read_counts_PlatesGH %>%
    tidyr::separate(file, into = c(NA, "gpfs", "home", "b042", "greenland", "plates", "platessub", "BWA", "sample", "fastqfile"), sep = "/", remove = TRUE) %>%
    mutate_at("sample", str_remove, "Sample_PRO1747_") %>% # remove index sequence from end of string, Sample_PRO1747_
    dplyr::select(sample, tot_num_seqs = num_seqs) %>%
    dplyr::group_by(sample) %>%
    summarise(tot_num_seqs = sum(tot_num_seqs))

fastq_read_counts_PlatesABA2B2EFGH <- bind_rows(fastq_read_counts_PlatesAB, fastq_read_counts_PlatesA2B2, fastq_read_counts_PlatesEF, fastq_read_counts_PlatesGH)

# sanity check: should have: "IPO3916" "IPO3917" "PlateA"  "PlateB"  "PlateE"  "PlateF"  "PlateG"  "PlateH"
fastq_read_counts_PlatesABA2B2EFGH %>% 
    dplyr::select(sample) %>% 
    tidyr::separate(sample, sep = "_", into = c("sampleprefix", "samplesuffix")) %>%
    dplyr::count(sampleprefix) %>% 
    dplyr::arrange(sampleprefix)

rm(fastq_read_counts_PlatesAB, fastq_read_counts_PlatesA2B2, fastq_read_counts_PlatesEF, fastq_read_counts_PlatesGH)
```

```{r load input_data_step5_20190120.RData, eval=FALSE}
pathtoexampledata = "~/Dropbox/Working_docs/Roslin_Greenland/2018/manuscript_methods/EcoLetts_20190120/Rcode/Rdata" 

load(file.path(pathtoexampledata, "input_data_step5_20190120.RData")) # new version of data tables after re-running all steps of the pipeline, to confirm reproducibility and to make a cleaner dataset for uploading.

tibble(env_data_COI)
tibble(env_data_mitogenome)
tibble(env_design)
tibble(env_seq_depth)  # tot_num_seqs for those samples
tibble(mock_data_COI)
tibble(mock_data_mitogenome)
tibble(mock_design)
tibble(mock_seq_depth)  # tot_num_seqs for those samples
tibble(species)
tibble(spikes)

# to make excel workbook so that i can see the target dataframes easily
# write_tsv(env_data_COI, "env_data_COI.tsv")
# write_tsv(env_data_mitogenome, "env_data_mitogenome.tsv")
# write_tsv(env_design, "env_design.tsv")
# write_tsv(env_seq_depth, "env_seq_depth.tsv")  # tot_num_seqs for those samples
# write_tsv(mock_data_COI, "mock_data_COI.tsv")
# write_tsv(mock_data_mitogenome, "mock_data_mitogenome.tsv")
# write_tsv(mock_design, "mock_design.tsv")
# write_tsv(mock_seq_depth, "mock_seq_depth.tsv")  # tot_num_seqs for those samples
# write_tsv(species, "species.tsv")
# write_tsv(spikes, "spikes.tsv")
```

```{r pathto20190115data}
pathto20190115data = "~/Dropbox/Working_docs/Roslin_Greenland/2017/bulk_samples/PlatesABA2B2EFGH"
```


```{r soupdata308}
# mapping outputs 
soupdata <- read_tsv(file.path(pathto20190115data, "minimap2_308mitogenomes_20190115/idx_meta_genomecov_ABA2B2EFGH_minimap2_20190115_308mitogenomes.txt.zip"))

## sanity check:  should have: "IPO3916" "IPO3917" "PlateA"  "PlateB"  "PlateE"  "PlateF"  "PlateG"  "PlateH"
soupdata %>% dplyr::select(sample) %>% tidyr::separate(sample, sep = "_", into = c("sampleprefix", "samplesuffix")) %>% dplyr::count(sampleprefix) %>% arrange(sampleprefix)

names(soupdata)

#  [1] "EI_RUN"                                      "Date"                                       
#  [3] "Plot"                                        "Trap"                                       
#  [5] "Week"                                        "mitogenome"                                 
#  [7] "Order"                                       "Family"                                     
#  [9] "Genus"                                       "Species_BOLD"                               
# [11] "mapped_reads"                                "mapped_reads_COI_corr"                      
# [13] "mapped_reads_COI_lysis_corr"                 "sum_coverage"                               
# [15] "mean_coverage"                               "stddev"                                     
# [17] "coefvar"                                     "pct_coverage"                               
# [19] "samtools_filter"                             "mt_length"                                  
# [21] "lysis_buffer_orig_total_ul"                  "lysis_buffer_purified_ul"                   
# [23] "lysis_buffer_proportion"                     "Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI"
# [25] "Coleoptera_Mordellidae_20ng_COI"             "Coleoptera_Elateridae_40ng_COI"             
# [27] "COI_corr_components"                         "COI_corr"                                   
# [29] "Concentration_ng_per_ul"                     "Sample_volume_ul"                           
# [31] "x.260_280_ratio"                             "sample"                                     
# [33] "ArcDyn_Plate_name"                           "EI_Plate_name"                              
# [35] "well"                                        "Sample_alias"                               
# [37] "TGAC_barcode"                                "pathname"                                   
# [39] "pathname_genomecov"                          "Full_name_of_the_sample"                    
# [41] "DateTrapEIRUN"                              



# mitogenomes_data
mitogenomes_data <- soupdata %>% dplyr::select(mitogenome, mt_length, Order, Family, Genus, Species_BOLD)

# env_design
## env_design, select sample, Date,
env_design <- soupdata %>% dplyr::select(sample = DateTrapEIRUN, date = Date, trap = Trap, run = EI_RUN, lysis_buffer_orig_total_ul, lysis_buffer_purified_ul, lysis_buffer_proportion)

## spike input DNA amount
## Plates AB & A2B2:  10, 20, 40 ng;  Plates EF & GH:  0.2, 0.4, 0.8 ng, for Bombycidae, Mordellidae, Elateridae, respectively
env_design <- env_design %>% mutate(
    Lepidoptera_Bombycidae_Bombyx_mori_COI = case_when(
        run == "idx_meta_genomecov_AB" ~ 10,
        run == "idx_meta_genomecov_A2B2" ~ 10,
        run == "idx_meta_genomecov_EF" ~ 0.2,
        run == "idx_meta_genomecov_GH" ~ 0.2
    )
)
env_design <- env_design %>% mutate(
    Coleoptera_Mordellidae_COI = case_when(
        run == "idx_meta_genomecov_AB" ~ 20,
        run == "idx_meta_genomecov_A2B2" ~ 20,
        run == "idx_meta_genomecov_EF" ~ 0.4,
        run == "idx_meta_genomecov_GH" ~ 0.4
    )
)
env_design <- env_design %>% mutate(
    Coleoptera_Elateridae_COI = case_when(
        run == "idx_meta_genomecov_AB" ~ 40,
        run == "idx_meta_genomecov_A2B2" ~ 40,
        run == "idx_meta_genomecov_EF" ~ 0.8,
        run == "idx_meta_genomecov_GH" ~ 0.8
    )
)

# summarise env_design
# env_design_summarised <- env_design %>%
#     dplyr::group_by(sample) %>%
#     dplyr::summarise(
#         date = first(date),
#         trap = first(trap),
#         run = first(run),
#         lysis_buffer_orig_total_ul = first(lysis_buffer_orig_total_ul),
#         lysis_buffer_purified_ul = first(lysis_buffer_purified_ul),
#         lysis_buffer_proportion = first(lysis_buffer_proportion),
#         Lepidoptera_Bombycidae_Bombyx_mori_COI = first(Lepidoptera_Bombycidae_Bombyx_mori_COI),
#         Coleoptera_Mordellidae_COI = first(Coleoptera_Mordellidae_COI),
#         Coleoptera_Elateridae_COI = first(Coleoptera_Elateridae_COI)
#     )

# env_data
## make long-format dataset of the COI_spike mapped reads, one row per sample and spike_species
## change names of the COI spike mapped reads to omit the ng portion
env_spikes <- soupdata %>%
    dplyr::select(sample = DateTrapEIRUN,
                  Lepidoptera_Bombycidae_Bombyx_mori_COI = Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI,
                  Coleoptera_Mordellidae_COI = Coleoptera_Mordellidae_20ng_COI,
                  Coleoptera_Elateridae_COI = Coleoptera_Elateridae_40ng_COI,
                  sampleorig = sample) %>%
    tidyr::gather(species, mapped_reads_mitogenome,
                  Lepidoptera_Bombycidae_Bombyx_mori_COI,
                  Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    dplyr::distinct(sample, species, mapped_reads_mitogenome, sampleorig)

# extract mapped reads and pct_coverage for each mitogenome in each sample
env_data <- soupdata %>%
    dplyr::select(sample = DateTrapEIRUN, species = mitogenome, mapped_reads_mitogenome = mapped_reads, pct_coverage_mitogenome = pct_coverage, sampleorig = sample)

# bind rows with env_spikes dataset, with different cols for mapped reads to mitogenomes and COI spike
env_data <- bind_rows(env_data, env_spikes) %>% arrange(sample)

# add fastq read counts to dataset,
env_data <- left_join(env_data, fastq_read_counts_PlatesABA2B2EFGH, by = c("sampleorig" = "sample")) %>% dplyr::select(-sampleorig)

# create a column indicating mitogenome or COI_spike species, change column order
env_data$mito_spike <- if_else(grepl("_COI$", env_data$species), "spike", "mitogenome")
env_data <- env_data %>% dplyr::select(sample, mito_spike, everything())

# save(mitogenomes_data, env_design, env_data, file = "~/Dropbox/Working_docs/Roslin_Greenland/2018/manuscript_methods/20190109/Rcode/Rdata/arcdyn_suppfiles_soups_20190115_308mitogenomes.RData")
# loadfile syntax
# load("~/Dropbox/Working_docs/Roslin_Greenland/2018/manuscript_methods/20190109/Rcode/Rdata/arcdyn_suppfiles_soups_20190115_308mitogenomes.RData")
```


```{r mock data files for Otso on 20181221 using 308 mitogenomes}
# mock datasets 
setwd("~/Dropbox/Working_docs/Roslin_Greenland/2017/bulk_samples/PlatesABA2B2EFGH") 

mockEF <- read_excel("minimap2_308mitogenomes_20190115/mocks_idx_meta_genomecov_PlatesEF_20190115_308mitogenomes.xlsx", sheet = "mocks_idx_meta_genomecov_Plates", na = "NA")
mockEF <- hablar::convert(mockEF, num(Replicate, inputDNA_ng))
# mockEF$Replicate <- as.character(mockEF$Replicate)
mockGH <- read_excel("minimap2_308mitogenomes_20190115/mocks_idx_meta_genomecov_PlatesGH_20190115_308mitogenomes.xlsx", sheet = "mocks_idx_meta_genomecov_Plates", na = "NA")
mockGH <- hablar::convert(mockGH, num(Replicate, inputDNA_ng))

# old location
# mockEF <- read_excel("PlatesEF/EF_outputs_20190115_F2308_f0x2_q48_minimap2_308mitogenomes/mocks_idx_meta_genomecov_PlatesEF_20190115_308mitogenomes.xlsx", sheet = "mocks_idx_meta_genomecov_Plates", na = "NA")
# mockEF$Replicate <- as.character(mockEF$Replicate)
# mockGH <- read_excel("PlatesGH/GH_outputs_20190115_F2308_f0x2_q48_minimap2_308mitogenomes/mocks_idx_meta_genomecov_PlatesGH_20190115_308mitogenomes.xlsx", sheet = "mocks_idx_meta_genomecov_Plates", na = "NA")

mockall <- dplyr::bind_rows(mockEF, mockGH)

# add fastq read counts to dataset,
mockall <- left_join(mockall, fastq_read_counts_PlatesABA2B2EFGH, by = c("sample" = "sample"))

## to make mock_inputspp
### extract COI_spike columns, make a long-format dataset (with gather), change column name to mitogenome, reduce to unique COI_spike names, and add a new column with index numbers (1,2,3). Change COI_spike colnames to omit 10/20/40ng from name
mock_spikes <- mockall %>%
    dplyr::select(Lepidoptera_Bombycidae_Bombyx_mori_COI = Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI, Coleoptera_Mordellidae_COI = Coleoptera_Mordellidae_20ng_COI, Coleoptera_Elateridae_COI = Coleoptera_Elateridae_40ng_COI) %>%
    tidyr::gather(species, mapped_reads, Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    dplyr::select(mitogenome = species) %>%
    dplyr::distinct(mitogenome)

mock_spikes <- cbind(mock_spikes, order = seq(nrow(mock_spikes)))

### extract mitogenome column, reduce to unique mitogenome names, and add a new column with index numbers (1,...,308)
mock_mitogenomes <- mockall %>%
    dplyr::select(mitogenome, mt_length) %>%
    dplyr::distinct(mitogenome, mt_length)

mock_mitogenomes <- cbind(mock_mitogenomes, order = seq(nrow(mock_mitogenomes)))

### bind mock_mitogenomes and mock_spikes, add spike or sp as prefix, merge with index number, change column order
mock_inputspp <- dplyr::bind_rows(mock_mitogenomes, mock_spikes)
mock_inputspp$species <- if_else(grepl("_COI$", mock_inputspp$mitogenome), "spike", "sp")
mock_inputspp <- mock_inputspp %>%
    unite(species, c("species", "order"), sep = "") %>%
    dplyr::select(species, mitogenome, mt_length)


## to make mock_design
## sample	experiment	run	lysis	input species	input mount
### select columns except for the COI_spike columns
mock_design <- mockall %>%
    dplyr::select(plate = Plate, shortname = Expt_shortname, experiment, run = Replicate, mitogenome_spike = mitogenome, input_amount = inputDNA_ng)
### left_join mock_inputspp. Do not include lysis (not necessary), do include mitogenome name
mock_design <- dplyr::left_join(mock_design, mock_inputspp, by = c("mitogenome_spike" = "mitogenome")) %>%
    dplyr::select(plate, shortname, experiment, run, mitogenome_spike, input_species = species, input_amount)

### select COI_spike species
mock_spikes <- mockall %>% dplyr::select(plate = Plate, shortname = Expt_shortname, experiment, run = Replicate, Lepidoptera_Bombycidae_Bombyx_mori_COI = Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI, Coleoptera_Mordellidae_COI = Coleoptera_Mordellidae_20ng_COI, Coleoptera_Elateridae_COI = Coleoptera_Elateridae_40ng_COI) %>%
    tidyr::gather(mitogenome_spike, mapped_reads_COI, Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    dplyr::distinct(plate, shortname, experiment, run, mitogenome_spike, mapped_reads_COI)

### add input DNA amounts (since only PlatesEF & GH, the amts are 0.2, 0.4, 0.8 only)
mock_spikes <- mock_spikes %>% mutate(
    input_amount = case_when(
        mitogenome_spike == "Lepidoptera_Bombycidae_Bombyx_mori_COI" ~ 0.2,
        mitogenome_spike == "Coleoptera_Mordellidae_COI" ~ 0.4,
        mitogenome_spike == "Coleoptera_Elateridae_COI" ~ 0.8
    )
)

mock_spikes <- left_join(mock_spikes, mock_inputspp, by = c("mitogenome_spike" = "mitogenome")) # %>% tidyr::separate(sample, c(NA, "Run"), sep = "_", remove = FALSE) #

mock_design$input_amount <- as.numeric(mock_design$input_amount)
mock_spikes <- mock_spikes %>% dplyr::select(plate, shortname, experiment, run, mitogenome_spike, input_species = species, input_amount)

mock_design <- bind_rows(mock_design, mock_spikes) %>% arrange(plate, shortname, run) # 4976 obs
mock_design <- mock_design %>% unite("sample", c(plate, shortname, run), remove = FALSE)
names(mock_design)


## to make mock_data
## sample mitogenome_spike mapped_reads_mitogenome pct_coverage_mitogenome mapped_reads_COI

### select columns except for the COI_spike columns
mock_data <- mockall %>% dplyr::select(plate = Plate, shortname = Expt_shortname, experiment, run = Replicate, sample, mitogenome_spike = mitogenome, mapped_reads_mitogenome = mapped_reads, pct_coverage_mitogenome = pct_coverage)

mock_spikes <- mockall %>% dplyr::select(plate = Plate, shortname = Expt_shortname, experiment, run = Replicate, sample, Lepidoptera_Bombycidae_Bombyx_mori_COI = Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI, Coleoptera_Mordellidae_COI = Coleoptera_Mordellidae_20ng_COI, Coleoptera_Elateridae_COI = Coleoptera_Elateridae_40ng_COI, tot_num_seqs) %>%
    tidyr::gather(mitogenome_spike, mapped_reads_mitogenome, Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    dplyr::distinct(plate, shortname, experiment, run, sample, mitogenome_spike, mapped_reads_mitogenome)

# names(mock_data)
# names(mock_spikes)

# bind mock_data and mock_spikes
mock_data <- bind_rows(mock_data, mock_spikes) %>% arrange(plate, shortname, run)

# add fastq read counts to dataset,
mock_data <- left_join(mock_data, fastq_read_counts_PlatesABA2B2EFGH, by = c("sample" = "sample")) %>% dplyr::select(-sample) # remove sample column with Plate_well info

mock_data <- mock_data %>% unite("sample", c(plate, shortname, run), remove = FALSE) # create new sample column using (plate, shortname, run). NB R automatically replaces old sample col with this new sample col
names(mock_data)


## save data files
# save(mock_inputspp, mock_design, mock_data, file = "~/Dropbox/Working_docs/Roslin_Greenland/2018/manuscript_methods/20190109/Rcode/Rdata/arcdyn_suppfiles_mocks_20190115_308mitogenomes.RData")

# loadfile syntax
# load("~/Dropbox/Working_docs/Roslin_Greenland/2018/manuscript_methods/20190109/Rcode/Rdata/arcdyn_suppfiles_mocks_20190115_308mitogenomes.RData")
```


```{r sample data files for Otso on 20181221 using 406 COI barcodes}
setwd("~/Dropbox/Working_docs/Roslin_Greenland/2017/bulk_samples/PlatesABA2B2EFGH") 
getwd()

# COI_barcode soups dataset
soupdata <- read_tsv("minimap2_406barcodes_20190116/idx_meta_genomecov_ABA2B2EFGH_minimap2_20190116_406barcodes.txt.zip")
# soupdata <- read_tsv("send_to_ArcDyn_collaborators_20180727_COI_Barcodes_only/idx_meta_genomecov_ABA2B2EFGH_minimap2_COIBarcodesOnly_20180727.txt.zip")

# mitogenomes_data
mitogenomes_data <- soupdata %>% dplyr::select(mitogenome, mt_length, Order, Family, Genus, Species_BOLD)

# env_design
## env_design, select sample, Date,
env_design <- soupdata %>% dplyr::select(sample = DateTrapEIRUN, date = Date, trap = Trap, run = EI_RUN, lysis_buffer_orig_total_ul, lysis_buffer_purified_ul, lysis_buffer_proportion)

## spike input DNA amount
## Plates AB & A2B2:  10, 20, 40 ng;  Plates EF & GH:  0.2, 0.4, 0.8 ng, for Bombycidae, Mordellidae, Elateridae, respectively
env_design <- env_design %>% mutate(
    Lepidoptera_Bombycidae_Bombyx_mori_COI = case_when(
        run == "idx_meta_genomecov_AB" ~ 10,
        run == "idx_meta_genomecov_A2B2" ~ 10,
        run == "idx_meta_genomecov_EF" ~ 0.2,
        run == "idx_meta_genomecov_GH" ~ 0.2
    )
)
env_design <- env_design %>% mutate(
    Coleoptera_Mordellidae_COI = case_when(
        run == "idx_meta_genomecov_AB" ~ 20,
        run == "idx_meta_genomecov_A2B2" ~ 20,
        run == "idx_meta_genomecov_EF" ~ 0.4,
        run == "idx_meta_genomecov_GH" ~ 0.4
    )
)
env_design <- env_design %>% mutate(
    Coleoptera_Elateridae_COI = case_when(
        run == "idx_meta_genomecov_AB" ~ 40,
        run == "idx_meta_genomecov_A2B2" ~ 40,
        run == "idx_meta_genomecov_EF" ~ 0.8,
        run == "idx_meta_genomecov_GH" ~ 0.8
    )
)

names(env_design)
# summarise env_design
# env_design_summarised <- env_design %>%
#     dplyr::group_by(sample) %>%
#     dplyr::summarise(
#         date = first(date),
#         trap = first(trap),
#         run = first(run),
#         lysis_buffer_orig_total_ul = first(lysis_buffer_orig_total_ul),
#         lysis_buffer_purified_ul = first(lysis_buffer_purified_ul),
#         lysis_buffer_proportion = first(lysis_buffer_proportion),
#         Lepidoptera_Bombycidae_Bombyx_mori_COI = first(Lepidoptera_Bombycidae_Bombyx_mori_COI),
#         Coleoptera_Mordellidae_COI = first(Coleoptera_Mordellidae_COI),
#         Coleoptera_Elateridae_COI = first(Coleoptera_Elateridae_COI)
#     )

names(soupdata)
# env_data
## make long-format dataset of the COI_spike mapped reads, one row per sample and spike_species
## change names of the COI spike mapped reads to omit the ng portion
env_spikes <- soupdata %>% dplyr::select(sample = DateTrapEIRUN, Lepidoptera_Bombycidae_Bombyx_mori_COI = Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI, Coleoptera_Mordellidae_COI = Coleoptera_Mordellidae_20ng_COI, Coleoptera_Elateridae_COI = Coleoptera_Elateridae_40ng_COI, sampleorig = sample) %>%
    tidyr::gather(species, mapped_reads_mitogenome, Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    dplyr::distinct(sample, species, mapped_reads_mitogenome, sampleorig)

# extract mapped reads and pct_coverage for each mitogenome in each sample
env_data <- soupdata %>%
    dplyr::select(sample = DateTrapEIRUN, species = mitogenome, mapped_reads_mitogenome = mapped_reads, pct_coverage_mitogenome = pct_coverage, sampleorig = sample)

# bind rows with env_spikes dataset, with different cols for mapped reads to mitogenomes and COI spike
env_data <- bind_rows(env_data, env_spikes) %>% arrange(sample)

# add fastq read counts to dataset,
env_data <- left_join(env_data, fastq_read_counts_PlatesABA2B2EFGH, by = c("sampleorig" = "sample")) %>% dplyr::select(-sampleorig)

# create a column indicating mitogenome or COI_spike species, change column order
env_data$mito_spike <- if_else(grepl("_COI$", env_data$species), "spike", "mitogenome")
env_data <- env_data %>% dplyr::select(sample, mito_spike, everything())

names(env_data)

# save(mitogenomes_data, env_design, env_data, file = "~/Dropbox/Working_docs/Roslin_Greenland/2018/manuscript_methods/20190109/Rcode/Rdata/arcdyn_suppfiles_soups_20190116_406barcodes.RData")

```


```{r mock data files for Otso on 20181221 using 406 COI barcodes}
# mock datasets
## mockall
### read EF and GH mock datasets, select the same columns and bind_rows
setwd("~/Dropbox/Working_docs/Roslin_Greenland/2017/bulk_samples/PlatesABA2B2EFGH")


mockEF <- read_excel("minimap2_406barcodes_20190116/mocks_idx_meta_genomecov_PlatesEF_20190116_406barcodes.xlsx", sheet = "mocks_idx_meta_genomecov_Plates", na = "NA")
mockEF <- hablar::convert(mockEF, num(Replicate, inputDNA_ng))
mockGH <- read_excel("minimap2_406barcodes_20190116/mocks_idx_meta_genomecov_PlatesGH_20190116_406barcodes.xlsx", sheet = "mocks_idx_meta_genomecov_Plates", na = "NA")
mockGH <- hablar::convert(mockGH, num(Replicate, inputDNA_ng))

# mockEF <- read_excel("PlatesEF/EF_outputs_20190116_F2308_f0x2_q48_minimap2_406barcodes/mocks_idx_meta_genomecov_PlatesEF_20190116_406barcodes.xlsx", sheet = "mocks_idx_meta_genomecov_Plates", na = "NA") 
# mockEF <- hablar::convert(mockEF, num(Replicate, inputDNA_ng))
# mockGH <- read_excel("PlatesGH/GH_outputs_20190116_F2308_f0x2_q48_minimap2_406barcodes/mocks_idx_meta_genomecov_PlatesGH_20190116_406barcodes.xlsx", sheet = "mocks_idx_meta_genomecov_Plates", na = "NA")
# mockGH <- hablar::convert(mockGH, num(Replicate, inputDNA_ng))

mockall <- bind_rows(mockEF, mockGH)
names(mockall)

mockall <- mockall %>% dplyr::select(plate = Plate, shortname = Expt_shortname, experiment, run = Replicate, sample, mitogenome_spike = mitogenome, length = mt_length, mapped_reads_mitogenome = mapped_reads, pct_coverage_mitogenome = pct_coverage, input_amount = inputDNA_ng, Lepidoptera_Bombycidae_Bombyx_mori_COI = Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI, Coleoptera_Mordellidae_COI = Coleoptera_Mordellidae_20ng_COI, Coleoptera_Elateridae_COI = Coleoptera_Elateridae_40ng_COI)
names(mockall)

## to make mock_inputspp
### extract COI_spike columns, make a long-format dataset (with gather), change column name to mitogenome, reduce to unique COI_spike names, and add a new column with index numbers (1,2,3). Change COI_spike colnames to omit 10/20/40ng from name
mock_spikes <- mockall %>%
    dplyr::select(Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    tidyr::gather(species, mapped_reads, Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    dplyr::select(mitogenome_spike = species) %>%
    dplyr::distinct(mitogenome_spike)
mock_spikes <- cbind(mock_spikes, order = seq(nrow(mock_spikes)))

### extract mitogenome column (which is the name of the COI barcode seqs), reduce to unique names, and add a new column with index numbers (1,...,406)
mock_mitogenomes <- mockall %>%
    dplyr::select(mitogenome_spike, length) %>%
    dplyr::distinct(mitogenome_spike, length)
mock_mitogenomes <- cbind(mock_mitogenomes, order = seq(nrow(mock_mitogenomes)))

### bind mock_mitogenomes and mock_spikes, add spike or sp as prefix, merge with index number, change column order
mock_inputspp <- dplyr::bind_rows(mock_mitogenomes, mock_spikes)
mock_inputspp$species <- if_else(grepl("_COI$", mock_inputspp$mitogenome_spike), "spike", "sp")
mock_inputspp <- mock_inputspp %>%
    unite(species, c("species", "order"), sep = "") %>%
    dplyr::select(species, mitogenome_spike, length)


## to make mock_design
## sample	experiment	run	lysis	input species	input mount
### select columns except for the COI_spike columns
mock_design <- mockall %>%
    dplyr::select(plate, shortname, experiment, run, mitogenome_spike, input_amount)
### left_join mock_inputspp. Do not include lysis (not necessary), do include mitogenome name
mock_design <- dplyr::left_join(mock_design, mock_inputspp, by = c("mitogenome_spike" = "mitogenome_spike")) %>%
    dplyr::select(plate, shortname, experiment, run, mitogenome_spike, input_species = species, input_amount)

### select COI_spike species
mock_spikes <- mockall %>% dplyr::select(plate, shortname, experiment, run, Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    tidyr::gather(mitogenome_spike, mapped_reads_COI, Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    dplyr::distinct(plate, shortname, experiment, run, mitogenome_spike, mapped_reads_COI)

### add input DNA amounts (since only PlatesEF & GH, the amts are 0.2, 0.4, 0.8 only)
mock_spikes <- mock_spikes %>% mutate(
    input_amount = case_when(
        mitogenome_spike == "Lepidoptera_Bombycidae_Bombyx_mori_COI" ~ 0.2,
        mitogenome_spike == "Coleoptera_Mordellidae_COI" ~ 0.4,
        mitogenome_spike == "Coleoptera_Elateridae_COI" ~ 0.8
    )
)

mock_spikes <- left_join(mock_spikes, mock_inputspp, by = c("mitogenome_spike" = "mitogenome_spike")) # %>% tidyr::separate(sample, c(NA, "Run"), sep = "_", remove = FALSE) #

mock_spikes <- mock_spikes %>% dplyr::select(plate, shortname, experiment, run, mitogenome_spike, input_species = species, input_amount)

mock_design <- bind_rows(mock_design, mock_spikes) %>% arrange(plate, shortname, run) # 6544 obs
mock_design <- mock_design %>% unite("sample", c(plate, shortname, run), remove = FALSE)
names(mock_design)



## to make mock_data
## sample mitogenome_spike mapped_reads_mitogenome pct_coverage_mitogenome mapped_reads_COI

### select columns except for the COI_spike columns
mock_data <- mockall %>% dplyr::select(plate, shortname, experiment, run, sample, mitogenome_spike, mapped_reads_mitogenome, pct_coverage_mitogenome)

mock_spikes <- mockall %>% dplyr::select(plate, shortname, experiment, run, sample,  Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    tidyr::gather(mitogenome_spike, mapped_reads_mitogenome, Lepidoptera_Bombycidae_Bombyx_mori_COI, Coleoptera_Mordellidae_COI, Coleoptera_Elateridae_COI) %>%
    dplyr::distinct(plate, shortname, experiment, run, sample, mitogenome_spike, mapped_reads_mitogenome)

names(mock_data)
names(mock_spikes)

# bind rows
mock_data <- bind_rows(mock_data, mock_spikes) %>% arrange(plate, shortname, run)

# add fastq read counts to dataset,
mock_data <- left_join(mock_data, fastq_read_counts_PlatesABA2B2EFGH, by = c("sample" = "sample")) %>% dplyr::select(-sample) # remove sample column with Plate_well info

mock_data <- mock_data %>% unite("sample", c(plate, shortname, run), remove = FALSE)

# create a column indicating mitogenome or COI_spike species, change column order
mock_data$mito_spike <- if_else(grepl("_COI$", mock_data$mitogenome_spike), "spike", "mitogenome")
mock_data <- mock_data %>% dplyr::select(sample, mito_spike, everything())

names(mock_data)

## save data files
# save(mock_inputspp, mock_design, mock_data, file = "~/Dropbox/Working_docs/Roslin_Greenland/2018/manuscript_methods/20190109/Rcode/Rdata/arcdyn_suppfiles_mocks_20190116_406barcodes.RData")

## load datafiles
# load("~/Dropbox/Working_docs/Roslin_Greenland/2018/manuscript_methods/20190109/Rcode/Rdata/arcdyn_suppfiles_mocks_20190116_406barcodes.RData")

```






