---
title: "idxstats_cbind"
author: "Douglas Yu"
date: "28/01/2018"
output: html_document
---

This file takes the idxstats.txt outputs of samtools and bedtools (idxstats, genomecov) and combines them with the sample metadata, does some sanity checks, removes failed samples, and makes it suitable for merging with the other Runs. This is for PlatesEF. 

This code has only been checked on macOS.

```{r}
# rm(list=ls())
library(tidyverse)
# ── Attaching packages ───────────────────────────────────────────────────────────────── tidyverse 1.2.1 ──
# ✔ ggplot2 3.1.0     ✔ purrr   0.2.5
# ✔ tibble  2.0.1     ✔ dplyr   0.7.8
# ✔ tidyr   0.8.2     ✔ stringr 1.3.1
# ✔ readr   1.3.1     ✔ forcats 0.3.0
# ── Conflicts ──────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
# ✖ dplyr::filter() masks stats::filter()
# ✖ dplyr::lag()    masks stats::lag()

library(readxl)
library(lubridate)
library(knitr)
library(beepr)
library(hablar)
sessionInfo()
```


```{r notify}
# function to allow macOS system notifications at end of a long command
notify <- function(msgString='Message from R', titleString='Message from R', speakIt=FALSE) {
    cmd <- paste('terminal-notifier -message ', '"', msgString, '"  -title "', titleString, '"', sep='')
    system(cmd)

    if (speakIt) {
        system(paste('say', msgString))
    }
}
```


```{r set paths and filenames for PlatesEF}
# setwd("~/Dropbox/Working_docs/Roslin_Greenland/2017/bulk_samples/PlatesEF")
setwd("../../PlatesEF") # go to folder with samtools/bedtools outputs
getwd()

samtoolsfilter <- "F2308_f0x2" # f0x2 is for PROPER PAIR, there are also samtools outputs for F2308 only
samtoolsqual <- "q48"

### use this version for 308 mitogenomes
# outputs_PlatesEF_F2308_f0x2_q48_minimap2_20180527
# idxstatsgenomecovfolder <- paste0("outputs_PlatesEF_", samtoolsfilter, "_", samtoolsqual, "_minimap2_outputs_20190115_308mitogenomes")  # this is the enclosing folder around the idxstats and genomcov files for a given set of soups
# outputs_PlatesEF_F2308_f0x2_q48_minimap2_outputs_20190115_308mitogenomes
# this is where i store the outputs of this script. Select according 
# outputidxstatstabulatefolder <- "EF_outputs_20190115_F2308_f0x2_q48_minimap2_308mitogenomes"

### use this version for 406 COI_barcodes
# idxstatsgenomecovfolder <- paste0("outputs_PlatesEF_", samtoolsfilter, "_", samtoolsqual, "_minimap2_outputs_20190116_406barcodes")  # this is the enclosing folder around the idxstats and genomcov files for a given set of soups
# outputs_PlatesEF_F2308_f0x2_q48_minimap2_outputs_20190116_406barcodes
# this is where i store the outputs of this script. Select according 
# barcodes
# outputidxstatstabulatefolder <- "EF_outputs_20190116_F2308_f0x2_q48_minimap2_406barcodes"

# use this version for 349 mitogenomes&COI_CYTB sequences
idxstatsgenomecovfolder <- paste0("outputs_PlatesEF_", samtoolsfilter, "_", samtoolsqual, "_minimap2_outputs_20190202_349MITOCOICYTB")  # this is the enclosing folder around the idxstats and genomcov files for a given set of soups
# outputs_PlatesEF_F2308_f0x2_q48_minimap2_outputs_20190202_349MITOCOICYTB
# this is where i store the outputs of this script
outputidxstatstabulatefolder <- "EF_outputs_20190202_F2308_f0x2_q48_minimap2_349MITOCOICYTB"

idxstatsfile <- paste0("*", samtoolsfilter, "_", samtoolsqual, "_sorted.bam_idxstats\\.txt")  
    # e.g. "*F2308_f0x2_q1_sorted.bam_idxstats\\.txt"
# choose the filename suffix to match the desired samtools filter. precede with * as a wildcard. \\ is used to escape the . before txt (because it is read with grep as "any character" in list.files() below)

genomecovfile <- paste0("*", samtoolsfilter, "_", samtoolsqual, "_sorted_genomecov_d\\.txt\\.gz")  
    # e.g. ""*F2308_f0x2_q1_sorted_genomecov_d\\.txt\\.gz""
# choose the filename pattern to match the desired samtools filter. precede with * as a wildcard for the sample names (e.g. Sample_IPO3916_A10_).      

samplemetadatafolder <- "~/Dropbox/Working_docs/Roslin_Greenland/2017/bulk_samples/ArcDyn_scripts/6_referenceseqs_metadata/PlatesEF_metadata"  # this is the enclosing folder around the sample metadata spreadsheet

samplemetadatafilename <- "Samples_for_sequencing_Zack_collection_Nov_2016_DY_TH_17_Feb_2017.xlsx"

lysisbufferfilename <- "lysis_buffer_EF.xlsx"  # this is based on Yinqiu's lysis buffer datasheet, which i edited a bit to make the sample names consistent so that tables could be joined via "Short_name_of_the_sample".  Also, i replaced missing data or cells with notes with NA
```


Read in all idx_stats files, add metadata to columns, and merge into one big table in long format (vertically by time)
```{r 1. read and tabulate idx files for PlatesEF}
# produces a list of unique idx filenames (full filenames but no path)
# use this command if i have a particular folder2 value (e.g. BWA02) to pull from
# idx_files <- list.files(file.path(idxstatsgenomecovfolder, folder2), pattern = "*F2308_f0x2_q1_sorted.bam_idxstats\\.txt")

cat("idxstatsgenomecoverfolder is: ", idxstatsgenomecovfolder)
cat("idxstatsfile search pattern is: ", idxstatsfile)

# idx_files <- list.files(Sys.glob(file.path(idxstatsgenomecovfolder, "BWA??", folder2)), pattern = idxstatsfile, full.names = TRUE) # old version
idx_files <- list.files(Sys.glob(file.path(idxstatsgenomecovfolder)), pattern = idxstatsfile, full.names = TRUE)
head(idx_files); cat("    ", length(idx_files), "total files")
    # Sys.glob is used to generate:  [1] "idxstats_files/BWA01" "idxstats_files/BWA02"
    # list.files is used to look in those folders for files matching "*F2308_f0x2_q1_sorted.bam_idxstats\\.txt"
    # full.names = TRUE is used to return the pathnames
    # \\ is used to escape the . before txt (because it is read with grep as "any character")

# column names of idx files 
idx_cols <- c("mitogenome", "mt_length", "mapped_reads", "unmapped_reads")

# function to read_tsv the file and filename and pathname
loadFile1 <- function(x) {
    # read in the four columns and store in df
    df <- read_tsv(x,   # previously file.path(x)
    col_names = idx_cols, na = "NA",
    col_types = cols(
        mitogenome = col_character(),
        mt_length = col_integer(),
        mapped_reads = col_integer(),
        unmapped_reads = col_integer()
        ))

    # read in filename, extract and sub in the first bit of the name, remove any filepaths (using basename) and store in df$sample.  example filename:  
    # PlateE_A1_AACCAACCG-GCCAAT_F2308_f0x2_q1_sorted.bam_idxstats.txt # PlatesEF

    df$sample <- sub("(Plate[A-Z]_\\D\\d+)_[A-Z-]+-[A-Z]+_F[\\w]+_sorted.bam_idxstats.txt", "\\1", basename(x), perl=TRUE)
    # [:alnum:] any alphanumeric character
    # \\w any letter, digit, underscore
    # \\D any non-digit, \\d any digit

    df$well <- sub("Plate[A-Z]_(\\D\\d+)_[A-Z-]+-[A-Z]+_F[\\w]+_sorted.bam_idxstats.txt", "\\1", basename(x), perl=TRUE)
    # extract well number from filename
    # \\w any letter, digit, underscore

    df$samtools_filter <- sub("Plate[A-Z]_\\D\\d+_[A-Z-]+-[A-Z]+_(F[\\w]+)_sorted.bam_idxstats.txt", "\\1", basename(x), perl=TRUE)
    # extract samtools parameters from filename

    df$pathname <- x
    # store full pathname in df$pathname

    # output is df
    df
}

# lapply the loadFile() function to all of the idx_files, save the output in idx as a list
idx <- purrr::map(idx_files, loadFile1) # map is equivalent to lapply

# Code from:  Location 6472 in Buffalo, Vince. Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools (Kindle Location 6493). O'Reilly Media. Kindle Edition.

# combine lists into a single dataframe (using do.call(rbind, idx)), reorder columns, delete unmapped_reads column, and remove rows that have "*" in the mitogenome field (the last line of each idxstats table)
idx <- do.call(rbind, idx) %>% 
    dplyr::select(mitogenome, mt_length, mapped_reads, sample, well, samtools_filter, pathname) %>%
    dplyr::filter(mitogenome != "*")

# sanity checks
idx %>% distinct(well) %>% count()  # 96, number of distinct values of well, as a sanity check
idx %>% distinct(sample) %>% count()  # number of distinct values of sample:  192  I originally input 192
idx %>% distinct(pathname) %>% count()  # number of distinct values of sample:  192
idx %>% distinct(samtools_filter) %>% count() # number of distinct values of samtools_filter: should be 1
idx %>% distinct(samtools_filter) # should look something like:  F2308_f0x2_q1
idx_mitogenome_table <- idx %>% group_by(well) %>% distinct(mitogenome) %>% count() %>% arrange(desc(n)); View(idx_mitogenome_table) # should be 311 in all rows (308 mitogenomes + 3 COIspikes); 409 (406 barcodes + 3 COI spikes); 349mitoCOICYTB: all cols should be 352 (349 + 3)
length(idx$mitogenome) / idx_mitogenome_table[1,2] == idx %>% distinct(sample) %>% count()  # should evaluate to TRUE
# e.g. 59712 rows / 311 mitogenomes == 192 samples;  78528 rows / 409 barcodes == 192
```


Read in sample metadata
```{r 2. read sample metadata and create Short_name_of_the_sample}
samplemetadata <- read_excel(file.path(samplemetadatafolder, samplemetadatafilename), sheet = "edited", na = "NA")

# use tidyr::separate to parse out metadata from Full_name_of_the_sample
samplemetadata <- samplemetadata %>% 
    tidyr::separate(Full_name_of_the_sample, c("Date", "Plot", "Trap", "Week"), sep = "_", remove = FALSE)
# Warning message:
# Expected 4 pieces. Missing pieces filled with `NA` in 6 rows [187, 188, 189, 190, 191, 192]. 
# these are the mock samples

samplemetadata$Date <- as_date(ymd(samplemetadata$Date))

samplemetadata$Year <- year(samplemetadata$Date)

samplemetadata <- samplemetadata %>% 
    dplyr::select(Full_name_of_the_sample, Date, Year, Plot:Extraction_kit_protocol)

# use tidyr::unite to make Short_name_of_the_sample
samplemetadata <- samplemetadata %>% 
    tidyr::unite("Short_name_of_the_sample", c("Year", "Plot", "Trap", "Week"))
```


```{r 2. read lysis buffer info and join to sample metadata}
lysisbufferdata <- read_excel(file.path(samplemetadatafolder, lysisbufferfilename), sheet = "lysis_buffer_EF", na = "NA")

lysisbufferdata <- lysisbufferdata %>%
    dplyr::mutate(lysis_buffer_orig_total_ul=lysis_buffer_orig_total_ml*1000)

lysisbufferdata <- lysisbufferdata %>%
    dplyr::mutate(lysis_buffer_purified_ul=lysis_buffer_purified_ml*1000)

lysisbufferdata <- lysisbufferdata %>% 
    dplyr::select(Short_name_of_the_sample, lysis_buffer_orig_total_ul, lysis_buffer_purified_ul)

samplemetadata <- left_join(samplemetadata, lysisbufferdata) # R message: "Joining, by = "Short_name_of_the_sample""

samplemetadata <- samplemetadata %>%
    dplyr::mutate(lysis_buffer_proportion=lysis_buffer_orig_total_ul/lysis_buffer_purified_ul)

samplemetadata$lysis_buffer_proportion <- round(samplemetadata$lysis_buffer_proportion, 2) # round to 2 decimal places
```


```{r 3. left_join sample metadata to idx table}
samplemetadata$sample <- str_c("Plate", substr(samplemetadata$ArcDyn_Plate_name, 1, 1), "_", samplemetadata$well)
    # substr(samplemetadata$ArcDyn_Plate_name, 1, 1) # extract first character 
    # str_c # concatenate multiple strings into a single string
    # alternative for substr is tidyr::str_sub(samplemetadata$ArcDyn_Plate_name, 1, 1)

idx_meta <- left_join(idx, samplemetadata) # R message:  "Joining, by = c("sample", "well")"

# use tidyr::separate to parse out metadata from Short_name_of_the_sample
idx_meta <- idx_meta %>% 
    tidyr::separate(Short_name_of_the_sample, c("Year", "Plot", "Trap", "Week"), sep = "_", remove = TRUE)

# sort columns, remove useless columns
idx_meta <- idx_meta %>% 
    dplyr::select(Date, Plot, Trap, Week, everything()) %>% 
    dplyr::select(-sample_reference, -species_name, -barcode_used, -Sample_type, -x.260_230_ratio, -x.260_280_ratio, -Quantification_method, -Buffer_used, -pathname, -Extraction_kit_protocol) %>% 
    dplyr::arrange(Date, Trap) 

names(idx_meta) 
#  [1] "Date"                       "Plot"                       "Trap"                      
#  [4] "Week"                       "mitogenome"                 "mt_length"                 
#  [7] "mapped_reads"               "sample"                     "well"                      
# [10] "samtools_filter"            "Full_name_of_the_sample"    "Year"                      
# [13] "Sample_alias"               "ArcDyn_Plate_name"          "EI_Plate_name"             
# [16] "DNA_conc"                   "Concentration_ng_per_ul"    "Sample_volume_ul"          
# [19] "lysis_buffer_orig_total_ul" "lysis_buffer_purified_ul"   "lysis_buffer_proportion"   
```


```{r sanity checks}
idx_meta %>% distinct(Date) %>% count()  # number of distinct values of Date:  72
idx_meta %>% distinct(Plot) %>% count()  # number of distinct values of Plot:  2 because we also have the mocks
idx_meta %>% distinct(Trap) %>% count()  # number of distinct values of Trap:  should be 3 (A,B,C) but we also have the mocks
idx_meta %>% distinct(Week) %>% count()  # number of distinct values of Week:  16
idx_meta %>% distinct(mitogenome) %>% count()  # number of distinct values of mitogenome:  311 (308 mitogenomes + 3 COI_spike);  barcodes:  409 (406 + 3 COI spikes); 349mitoCOICYTB:  352 (349 + 3)
idx_meta %>% distinct(sample) %>% count()  # number of distinct values of sample:   192
idx_meta %>% distinct(sample)
idx_meta %>% distinct(Sample_alias) %>% count()  # number of distinct values of Sample_alias:  should be = sample = 192
idx_meta %>% distinct(Sample_alias)
idx_meta %>% distinct(well) %>% count()  # number of distinct values of well:  96
idx_meta %>% distinct(ArcDyn_Plate_name) %>% count()  # number of distinct values of ArcDyn_Plate_name:  should be 2
idx_meta %>% distinct(samtools_filter) %>% count()  # number of distinct values of samtools_filter: should be 1
idx_meta %>% distinct(Full_name_of_the_sample) %>% count()  # number of distinct values of Full_name_of_the_sample: should be 192 
idx_meta %>% distinct(Full_name_of_the_sample) %>% arrange(Full_name_of_the_sample) # number of distinct values of samtools_filter: should be 192
```

In the next code chunk, it takes 30-40 mins to read in all the genomecov_d.txt.gz files (used for calculating percent_coverage). This only needs to happen once, and I have saved the table after input. Skip to halfway down the chunk to "TIME SAVING STEP" where there are two readRDS() commands for reading these in, one for mitogenomes, one for barcodes. 

```{r 4. read in genomecov_d.txt.gz files and add to idx_meta}
# genomecovfile <- "*F2308_f0x2_q1_sorted_genomecov_d\\.txt\\.gz"  # choose the filename pattern to match the desired samtools filter. precede with * as a wildcard for the sample names (e.g. Sample_IPO3916_A10_).      
# \\ is used to escape the . before txt (because it is read with grep as "any character" in list.files()

cat("Pattern search is for: ", genomecovfile)

genomecov_files <- list.files(Sys.glob(file.path(idxstatsgenomecovfolder)), pattern = genomecovfile, full.names = TRUE)
head(genomecov_files); cat("    ", length(genomecov_files), "total files")
    # Sys.glob is used to generate:  
        # [1] "minimap2_all_outputs_PlatesAB/Sample_IPO3916_A1_F2308_f0x2_q1_sorted_genomecov_d.txt.gz"
    # list.files is used to look in those folders for files matching "*F2308_f0x2_q1_sorted_genomecov_d\\.txt\\.gz"
    # full.names = TRUE is used to return the pathnames
    # \\ is used to escape the . before txt and gz (because it is read with grep as "any character" in list.files()

# column names of genomecov_files
columnnames <- c("mitogenome", "position", "coverage")

# function to read_tsv the file
loadFile2 <- function(x) {
    df <- read_tsv(gzfile(x),   
    # e.g. "minimap2_all_outputs_PlatesEF/PlateE_A1_AACCAACCG-GCCAAT_F2308_f0x2_q1_sorted_genomecov_d.txt.gz"
    col_names = columnnames, na = "NA",
    # col_types = cols(
    #     mitogenome = col_character(),
    #     position = col_integer(),
    #     coverage = col_double()),
    trim_ws = TRUE
    )

    # read in filename "basename(x)", extract the sample name "\\1" from  PlateE_A1_AACCAACCG-GCCAAT_F2308_f0x2_q1_sorted_genomecov_d.txt.gz, and place in df$sample column
    df$sample <- sub("(Plate[A-Z]_\\D\\d+)_[A-Z-]+-[A-Z]+_F[\\w]+_sorted_genomecov_d.txt.gz", "\\1", basename(x), perl=TRUE)
    # \\w any letter, digit, underscore
    # \\D any non-digit, \\d any digit

    df$well <- sub("Plate[A-Z]_(\\D\\d+)_[A-Z-]+-[A-Z]+_F[\\w]+_sorted_genomecov_d.txt.gz", "\\1", basename(x), perl=TRUE)
    # extract well number from filename
    # \\w any letter, digit, underscore
    # \\D any non-digit, \\d any digit
    
    df$samtools_filter <- sub("Plate[A-Z]_\\D\\d+_[A-Z-]+-[A-Z]+_(F[\\w]+)_sorted_genomecov_d.txt.gz", "\\1", basename(x), perl=TRUE)
    # extract samtools parameters from filename

    df$pathname_genomecov <- x
    # store full pathname in df$pathname

    df <- df %>% dplyr::group_by(mitogenome) %>% 
        summarise(sum_coverage = sum(coverage), 
                  mean_coverage = mean(coverage), 
                  stddev = sd(coverage), 
                  coefvar = sd(coverage)/mean(coverage), 
                  length = n(), 
                  pct_coverage = sum(coverage>0)/n(), # % of positions that have 1 or more reads mapped 
                  sample = first(sample), 
                  well = first(well), 
                  samtools_filter = first(samtools_filter), 
                  pathname_genomecov = first(pathname_genomecov)
                  )

    # output is df
    df
}


# this step requires ~30 minutes to run for mitogenomes, ~ 2 mins for barcodes
# lapply the loadFile2() function to all of the genomecov_files, save the output in genomecovfiles as a list
# map is equivalent to lapply
genomecoverages_summ <- purrr::map(genomecov_files, loadFile2); notify("Your loadfile2 function has finally finished!", "Message from R", speakIt=TRUE)

# rbind the list into a dataframe
genomecoverages_summ <- do.call(rbind, genomecoverages_summ)

# TIME SAVING STEP  After running once for mitogenomes and for barcodes, i save the file to avoid having to rerun the map() step
    # mitogenomes
    # genomecoverages_summ <- readRDS(file.path(outputidxstatstabulatefolder, "genomecoverages_summ_PlatesEF_20190115_308mitogenomes.RDS"))
    # barcodes
    # genomecoverages_summ <- readRDS(file.path(outputidxstatstabulatefolder, "genomecoverages_summ_PlatesEF_20190116_406barcodes.RDS"))
    # genomecoverages_summ <- readRDS(file.path(outputidxstatstabulatefolder, "genomecoverages_summ_PlatesEF_20190202_349MITOCOICYTB.RDS"), compress = TRUE)

    # the commands used to save the genomecov files
    # saveRDS(file.path(genomecoverages_summ, file.path(outputidxstatstabulatefolder, "genomecoverages_summ_PlatesEF_20190115_308mitogenomes.RDS"), compress = TRUE)
   # saveRDS(file.path(genomecoverages_summ, file.path(outputidxstatstabulatefolder, "genomecoverages_summ_PlatesEF_20190116_406barcodes.RDS"), compress = TRUE)
    # saveRDS(genomecoverages_summ, file.path(outputidxstatstabulatefolder, "genomecoverages_summ_PlatesEF_20190202_349MITOCOICYTB.RDS"), compress = TRUE)

# left_join genomecoverages_summ to idx_meta
idx_meta_genomecov <- left_join(idx_meta, genomecoverages_summ) # R message:  Joining, by = c("mitogenome", "sample", "well", "samtools_filter")

idx_meta_genomecov <- idx_meta_genomecov %>% arrange(Date, Trap)

# use grepl() to find mitogenome names that are COI spike names
idx_meta_genomecov$COI_Species <- if_else(grepl("_COI$", idx_meta_genomecov$mitogenome), "COI_Spike", "ArcDyn_Species") # this regex works for both the mitogenome and barcode datasets

# reorder the columns for convenience
idx_meta_genomecov <- idx_meta_genomecov %>% 
    dplyr::select(Date:mitogenome, COI_Species, mapped_reads, everything()) 

# change the names of the spikes from an old naming convention to the one that we want to use
# must do this at this stage, after joining idxstats and genomecov files
idx_meta_genomecov <-  idx_meta_genomecov %>% 
    mutate(
        mitogenome = case_when(
            mitogenome == "1-2_Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI" ~  "Lepidoptera_Bombycidae_Bombyx_mori_COI",
             mitogenome == "1-2_Lepidoptera_Bombycidae_Bombyx_mori_COI" ~ "Lepidoptera_Bombycidae_Bombyx_mori_COI",
           mitogenome == "3-1_Coleoptera_Mordellidae_20ng_COI" ~ "Coleoptera_Mordellidae_COI",
            mitogenome == "3-1_Coleoptera_Mordellidae_COI" ~ "Coleoptera_Mordellidae_COI",
            mitogenome == "2-1_Coleoptera_Elateridae_40ng_COI" ~ "Coleoptera_Elateridae_COI",
            mitogenome == "2-1_Coleoptera_Elateridae_COI" ~ "Coleoptera_Elateridae_COI",
            TRUE ~ as.character(mitogenome)
        )
    )

names(idx_meta_genomecov)
#  [1] "Date"                       "Plot"                       "Trap"                      
#  [4] "Week"                       "mitogenome"                 "COI_Species"               
#  [7] "mapped_reads"               "mt_length"                  "sample"                    
# [10] "well"                       "samtools_filter"            "Full_name_of_the_sample"   
# [13] "Year"                       "Sample_alias"               "ArcDyn_Plate_name"         
# [16] "EI_Plate_name"              "DNA_conc"                   "Concentration_ng_per_ul"   
# [19] "Sample_volume_ul"           "lysis_buffer_orig_total_ul" "lysis_buffer_purified_ul"  
# [22] "lysis_buffer_proportion"    "sum_coverage"               "mean_coverage"             
# [25] "stddev"                     "coefvar"                    "length"                    
# [28] "pct_coverage"               "pathname_genomecov"        
```


```{r cleanup files}
rm(genomecoverages_summ)
rm(idx)
rm(idx_meta)
rm(idx_mitogenome_table)
rm(lysisbufferdata)
rm(samplemetadata)
```


Filter out failed samples. There should always be mapped reads for the COI_spike species, so if this number is zero or very low, the sample has failed and should be deleted.
```{r}
# calculate sum of all COI spike reads for each sample. 
sum_idx_meta_genomecov <- idx_meta_genomecov %>%
    group_by(Full_name_of_the_sample, COI_Species) %>%
    summarise_at(vars(mapped_reads), funs(sum(as.numeric(.)))) %>%
    arrange(desc(COI_Species), mapped_reads, Full_name_of_the_sample)

# visual inspection reveals 1 sample that has very few COI_spike mapped reads (8). Remove it
sample_names <- sum_idx_meta_genomecov %>% 
    dplyr::filter(COI_Species == "COI_Spike" && mapped_reads < 200) %>% 
    dplyr::select(Full_name_of_the_sample) %>% 
    distinct(Full_name_of_the_sample)

samples_to_remove <- sample_names %>% 
    dplyr::pull(Full_name_of_the_sample) # pull out a single variable
samples_to_remove # 1 sample to remove
# [1] "2009Jun17_Art3_TrapB_Wk25"
idx_meta_genomecov <- idx_meta_genomecov %>% 
    filter(!Full_name_of_the_sample %in% samples_to_remove) # 59401 obs mitogenomes, 65849 obs barcodes, 67232 obs MITOCOICYTB

# sanity check:  191 samples, which is correct since i removed 1 sample
idx_meta_genomecov %>% 
    distinct(Full_name_of_the_sample) %>% 
    count()  # number of distinct values of Full_name_of_the_sample
```

Sanity checks and save idx_meta_genomecov output file for merging with the other RUN outputs. Also, separate out mock data for further formatting
```{r separate mock and environmental samples}
# confirm that mock and environmental samples are in the dataset
idx_meta_genomecov %>% 
    distinct(Plot) %>% 
    arrange(Plot) # distinct values of Plot: should be 2 (Art3, evenDNA). Art3 is the name of the Zackenberg plot, and evenDNA is the name of the mock even samples

# mock data table
mocks_idx_meta_genomecov <- idx_meta_genomecov %>% 
    dplyr::filter(!Plot == "Art3")

# remove mock soup data from idx_meta_genomecov
idx_meta_genomecov <- idx_meta_genomecov %>% 
    dplyr::filter(Plot == "Art3")

# create variable from Date and Trap(ABC)
# stringr::str_pad(x, width = 2, side = "left", pad = "0")  # to ensure that the date always has two digits, e.g. 07.
idx_meta_genomecov$DateTrap <- str_c("Date", year(idx_meta_genomecov$Date), str_pad(month(idx_meta_genomecov$Date), 2, side = "left", pad = "0"), str_pad(day(idx_meta_genomecov$Date), 2, side = "left", pad = "0"), idx_meta_genomecov$Trap, sep = "_")

idx_meta_genomecov <- idx_meta_genomecov %>% select(DateTrap, everything())

# in coefvar column, replace "NaN" string to "NA" and then coerce column back to numeric
idx_meta_genomecov$coefvar <- idx_meta_genomecov$coefvar %>%  str_replace("NaN", "NA") %>% as.numeric()

# these tables check that i only have one row for each Full_name_of_the_sample, DateTrap, and mitogenome
# ideally, there should be a 1 in all rows of column "n"
idxstats_by_Full_name_of_the_sample <- idx_meta_genomecov %>% 
    dplyr::group_by(DateTrap) %>% 
    dplyr::distinct(Full_name_of_the_sample) %>% 
    dplyr::count() %>% 
    dplyr::arrange(desc(n)); View(idxstats_by_Full_name_of_the_sample)

# ideally, there should be a 1 in all rows of column "n"
idxstats_by_DateTrap <- idx_meta_genomecov %>% 
    dplyr::group_by(Full_name_of_the_sample) %>% 
    dplyr::distinct(DateTrap) %>% 
    dplyr::count() %>% 
    dplyr::arrange(desc(n)); View(idxstats_by_DateTrap)

# ideally, there should be a 311, 409, or 352 in all rows of column "n"
idxstats_by_mitogenome <- idx_meta_genomecov %>% 
    dplyr::group_by(Full_name_of_the_sample) %>% 
    dplyr::distinct(mitogenome) %>% 
    dplyr::count() %>% 
    dplyr::arrange(desc(n)); View(idxstats_by_mitogenome)

# save idx_meta_genomecov files
# write_tsv(idx_meta_genomecov, file.path(outputidxstatstabulatefolder, "idx_meta_genomecov_PlatesEF_20190115_308mitogenomes.txt")) # output file for merging with the other RUN outputs
# write_tsv(idx_meta_genomecov, file.path(outputidxstatstabulatefolder, "idx_meta_genomecov_PlatesEF_20190116_406barcodes.txt")) # output file for merging with the other RUN outputs
# write_tsv(idx_meta_genomecov, file.path(outputidxstatstabulatefolder, "idx_meta_genomecov_PlatesEF_20190202_349MITOCOICYTB.txt")) # output file for merging with the other RUN outputs

# in case i want to read in the output datafiles
# idx_meta_genomecov <- read_delim(file.path(outputidxstatstabulatefolder, "idx_meta_genomecov_PlatesEF_20190115_308mitogenomes.txt"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)
# idx_meta_genomecov <- read_delim(file.path(outputidxstatstabulatefolder, "idx_meta_genomecov_PlatesEF_20190116_406barcodes.txt"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)
# idx_meta_genomecov <- read_delim(file.path(outputidxstatstabulatefolder, "idx_meta_genomecov_PlatesEF_20190202_349MITOCOICYTB.txt"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)
```


Clean up and save mock soup data
```{r}
# create columns 
mocks_idx_meta_genomecov$plate <- "PlateEF"

# create experiment column
mocks_idx_meta_genomecov <- mocks_idx_meta_genomecov %>% 
    mutate(
        experiment = case_when(
            Sample_alias == "ME50_1" | Sample_alias == "ME50_2" | Sample_alias == "ME100_1" | Sample_alias == "ME100_2" | Sample_alias == "ME200_1" | Sample_alias == "ME200_2" ~ "mockeven",
            Sample_alias == "MG_1" | Sample_alias == "MG_2" | Sample_alias == "MG_3" ~ "mockgradient",
            Sample_alias == "COI_1_negative_control" | Sample_alias == "COI_2_negative_control" ~ "negctrl"
        )
    )

# rename sample column to plate_well, because otherwise i lose this column in the next step
mocks_idx_meta_genomecov <- mocks_idx_meta_genomecov %>% 
    rename(plate_well = sample)  # orig colname is sample

# create sample column
mocks_idx_meta_genomecov <- mocks_idx_meta_genomecov %>% 
    tidyr::unite("sample", c("plate", "Sample_alias"), sep = "_")

# create plate, input_amount, run columns from Sample_alias column:  PlateEF_ME100_1
mocks_idx_meta_genomecov <- mocks_idx_meta_genomecov %>% 
    tidyr::separate(sample, c("plate", "input_amount", "run"), sep = "_", remove = FALSE, convert = TRUE)

# # remove ME substring from input_amount column and convert type to numeric
mocks_idx_meta_genomecov <- mocks_idx_meta_genomecov %>%
    mutate_at("input_amount", str_remove, "M[E,G]") %>% # remove ME or MG
    hablar::convert(num(input_amount))

# select and rename columns
mocks_idx_meta_genomecov <- mocks_idx_meta_genomecov %>% 
    dplyr::select(sample, plate, experiment, run, mitogenome, PC = pct_coverage, mapped_reads, input_amount, plate_well)

names(mocks_idx_meta_genomecov)
# [1] "sample"       "plate"        "experiment"   "run"          "mitogenome"   "PC"          
# [7] "mapped_reads" "input_amount" "plate_well"  



# save mocks_idx_meta_genomecov files

# write_tsv(mocks_idx_meta_genomecov, file.path(outputidxstatstabulatefolder, "mocks_idx_meta_genomecov_PlatesEF_20190115_308mitogenomes.txt")) # for merger with mock file from PlatesGH
# write_tsv(mocks_idx_meta_genomecov, file.path(outputidxstatstabulatefolder, "mocks_idx_meta_genomecov_PlatesEF_20190116_406barcodes.txt")) # for merger with mock file from PlatesGH
# write_tsv(mocks_idx_meta_genomecov, file.path(outputidxstatstabulatefolder, "mocks_idx_meta_genomecov_PlatesEF_20190202_349MITOCOICYTB.txt")) # for merger with mock file from PlatesGH

# read mocks_idx_meta_genomecov files
# mocks_idx_meta_genomecov <- read_delim(file.path(outputidxstatstabulatefolder, "mocks_idx_meta_genomecov_PlatesEF_20190115_308mitogenomes.txt"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)
# mocks_idx_meta_genomecov <- read_delim(file.path(outputidxstatstabulatefolder, "mocks_idx_meta_genomecov_PlatesEF_20190116_406barcodes.txt"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)
# mocks_idx_meta_genomecov <- read_delim(file.path(outputidxstatstabulatefolder, "mocks_idx_meta_genomecov_PlatesEF_20190202_349MITOCOICYTB.txt"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)
```


END




# cannot run this anymore because i no longer calculate COI_lysis_corrected_reads. These wide tables will need to be made with the new data tables
```{r Make wide tables for visualisation, eval = FALSE}
# Make wide table for visualisation (left to right should follow temporal order (DateTrap))
idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide <- idx_meta_genomecov %>% dplyr::select(DateTrap, mitogenome, mapped_reads_COI_lysis_corr) %>% dplyr::arrange(DateTrap) %>% tidyr::spread(DateTrap, mapped_reads_COI_lysis_corr) %>% dplyr::arrange(mitogenome)

# uncomment to save tables
# write_tsv(idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide, "idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide_PlatesEF_20190115_308mitogenomes.txt")
# write_tsv(idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide, "idx_meta_genomecov_mapped_reads_COI_lysis_corr_wide_PlatesEF_20190116_406barcodes.txt")

# make wide table with pct_coverage > pctcovmin > 0.1
plot(idx_meta_genomecov$pct_coverage, idx_meta_genomecov$coefvar)
# set a minimum value of coefvar, above which read numbers are set to -
hist(idx_meta_genomecov$pct_coverage, breaks = 2000, xlim = c(0.000, 1), ylim = c(0, 400)) # reasonable cutoff at 2.0? 3.0?
pctcovmin <- 0.10
idx_meta_genomecov$mapped_reads_COI_lysis_corr_pctcovmin <- ifelse(idx_meta_genomecov$pct_coverage >= pctcovmin, idx_meta_genomecov$mapped_reads_COI_lysis_corr, 0)
# change column order
idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::select(Date:mapped_reads_COI_lysis_corr, mapped_reads_COI_lysis_corr_pctcovmin, mean_coverage:DateTrap)

# # not necessary
# # replace NA values with 0s
# idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate_at(vars(mapped_reads_COI_lysis_corr_pctcovmin), funs(replace(., is.na(.), 0))) # the . stands in for the variables that are being assessed

# hist(idx_meta_genomecov$coefvar, breaks = 2000, xlim = c(0.0000, 5)) # reasonable cutoff at 2.0? 3.0?
# # set a minimum value of coefvar, above which read numbers are set to -
# coefvarmin <- 3.0
# idx_meta_genomecov$mapped_reads_COI_lysis_corr_coefvarmin <- ifelse(idx_meta_genomecov$coefvar <= coefvarmin, idx_meta_genomecov$mapped_reads_COI_lysis_corr, 0)
# # change column order
# idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::select(Date:mapped_reads_COI_lysis_corr, mapped_reads_COI_lysis_corr_coefvarmin, mean_coverage:DateTrap)
# # replace NA values with 0s
# idx_meta_genomecov <- idx_meta_genomecov %>% dplyr::mutate_at(vars(mapped_reads_COI_lysis_corr_coefvarmin), funs(replace(., is.na(.), 0))) # the . stands in for the variables that are being assessed

# mapped_reads_COI_lysis_corr_pctcovmin # read number = 0 if pct_coverage < pctcovmin (e.g. 0.10)
idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide <- idx_meta_genomecov %>% dplyr::select(DateTrap, mitogenome, mapped_reads_COI_lysis_corr_pctcovmin) %>% dplyr::arrange(DateTrap) %>% tidyr::spread(DateTrap, mapped_reads_COI_lysis_corr_pctcovmin) %>% dplyr::arrange(mitogenome)

# uncomment to save table
# write_tsv(idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide, "idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide_PlatesEF_20190115_308mitogenomes.txt")
# write_tsv(idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide, "idx_meta_genomecov_mapped_reads_COI_lysis_corr_pctcovmin_wide_PlatesEF_20190116_406barcodes.txt")


# Make wide table for visualisation of the mock soups
mocks_idx_meta_genomecov_mapped_reads_COI_corr_wide <- mocks_idx_meta_genomecov %>% dplyr::select(Sample_alias, mitogenome, mapped_reads_COI_corr) %>% dplyr::arrange(Sample_alias) %>% tidyr::spread(Sample_alias, mapped_reads_COI_corr) %>% dplyr::arrange(mitogenome)

mocks_idx_meta_genomecov_pct_coverage_wide <- mocks_idx_meta_genomecov %>% dplyr::select(Sample_alias, mitogenome, pct_coverage) %>% dplyr::arrange(Sample_alias) %>% tidyr::spread(Sample_alias, pct_coverage) %>% dplyr::arrange(mitogenome)

mocks_idx_meta_genomecov_mapped_reads_wide <- mocks_idx_meta_genomecov %>% dplyr::select(Sample_alias, mitogenome, mapped_reads) %>% dplyr::arrange(Sample_alias) %>% tidyr::spread(Sample_alias, mapped_reads) %>% dplyr::arrange(mitogenome)

# # uncomment to save tables (with and without COI_spike_correction)
# write_tsv(mocks_idx_meta_genomecov_mapped_reads_COI_corr_wide, "mocks_idx_meta_genomecov_mapped_reads_COI_corr_wide_PlatesEF_20190115_308mitogenomes.txt")
# write_tsv(mocks_idx_meta_genomecov_pct_coverage_wide, "mocks_idx_meta_genomecov_pct_coverage_wide_PlatesEF_20190115_308mitogenomes.txt")
# write_tsv(mocks_idx_meta_genomecov_mapped_reads_wide, "mocks_idx_meta_genomecov_mapped_reads_wide_PlatesEF_20190115_308mitogenomes.txt")#

# write_tsv(mocks_idx_meta_genomecov_mapped_reads_COI_corr_wide, "mocks_idx_meta_genomecov_mapped_reads_COI_corr_wide_PlatesEF_20190116_406barcodes.txt")
# write_tsv(mocks_idx_meta_genomecov_pct_coverage_wide, "mocks_idx_meta_genomecov_pct_coverage_wide_PlatesEF_20190116_406barcodes.txt")
# write_tsv(mocks_idx_meta_genomecov_mapped_reads_wide, "mocks_idx_meta_genomecov_mapped_reads_wide_PlatesEF_20190116_406barcodes.txt")#
```

```{r, eval=FALSE}
# purl("idxstats_tabulate_macOS.Rmd")
```


## Next steps:

1. DONE Read in all idx_stats files, add sample metadata to columns, and merge into one table in long format (vertically by time)
2. DONE merge with sample excel workbook to add the sample date information
3. DONE go through the genomecov files and calculate the mean and standard deviation of coverage for each mitogenome/sample and add the columns to the idx table
4. DONE make wide table using tidyr (the column data format should be DateTrap (e.g. 1998_08_05_TrapA) in temporal order).
5. DONE confirm that i have the correct number of samples (compare with number of input files) e.g. PlatesAB only have 176 bam files (or did i only download 176 bam files?) but of course we created 2*96=192 files, some of which failed to sequence
6. DONE Debug the bug that causes sample to have repeated lines, using a tabulate command to check that I don't have repeated lines.  Answer: re-run the minimap2/samtools/bedtools scripts and redownload the idxstats files.  i put the same idxstats files into multiple BWA folders in my test folders
7. DONE platesGH:  combine_fastq, fastQC, multiqc (using PKG-ENQ-2379-Data_Transfer-PSEQ-1586-trimmed), remove orig fastq.gz files.  
8. DONE platesA2B2: combine_fastq, fastQC, multiqc, remove orig fastq.gz files
9. DONE minimap2/samtools for platesGH
10. DONE minimap2/samtools for platesA2B2

11. create lysis buffer datasets, download idxstats and genomecov data, run R code on Plates AB, A2B2, EF, GH:  COMPARE PlatesAB and A2B2 WITH YINQIU'S BWA_results_20161025.xlsx.  Compare Plates?? with the positive controls, the seasonal patterns, the COI spikes

12. remove original, noncombined fastq.gz files from A2B2 and GH when i'm done mapping and checking.

13. make textfile for taxonomy (class, order, family, genus, species) of each mitogenome and join to idx_meta_genomecov

14. visualise minimap2 bam files and check that i am not mapping to the Ns between the protein-coding genes. SHOULD I SUBTRACT A FIXED NUMBER FROM EACH MITOGENOME LENGTH TO CALCULATE THE CORRECT LENGTH OF THE CODING PORTION?  






names(idx_meta_genomecov)
[1] "Date"                                        "Plot"
[3] "Trap"                                        "Week"
[5] "mitogenome"                                  "COI_Species"
[7] "mapped_reads"                                "mapped_reads_COI_corr"
[9] "mapped_reads_COI_lysis_corr"                 "mean_coverage"
[11] "stddev"                                      "coefvar"
[13] "length"                                      "sample"
[15] "well"                                        "samtools_filter"
[17] "pathname"                                    "Full_name_of_the_sample"
[19] "Sample_alias"                                "ArcDyn_Plate_name"
[21] "EI_Plate_name"                               "TGAC_barcode"
[23] "Concentration_ng_per_ul"                     "Sample_volume_ul"
[25] "Quantification_method"                       "x.260_280_ratio"
[27] "lysis_buffer_purified_ml"                    "lysis_buffer_orig_total_ml"
[29] "lysis_buffer_proportion"                     "sum_coverage"
[31] "pathname_genomecov"                          "Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI"
[33] "Coleoptera_Elateridae_40ng_COI"              "Coleoptera_Mordellidae_20ng_COI"
[35] "ratio_mordellid20_bombyx10"                  "ratio_elaterid40_bombyx"
[37] "ratio_elaterid40_mordellid20"                "elaterid_mordellid_sum"
[39] "COI_corr"

### Metadata:
From idxstats.txt
Date:  Date of ArcDyn Sample
Plot:  Zackenberg Plot (only Art3)
Trap:  Window trap within Art3 (A, B, C)
Week:  Week of the year (alternative to Date)
mitogenome:  name of mitogenome sequencer (from reference fasta file used by minimap2)
mt_length:  length of the mitogenome (including Ns between the genes) used by minimap2
mapped_reads:  number of reads mapped to each mitogenome by minimap2 and filtered through samtools
sample:  unique sample name derived from each sample's foldername (e.g. Sample_IPO3916_A1 -> IPO3916_A1). IPO3916 refers to the Plate and A1 refers to the well. I use this to link to my ArcDyn sample metadata
well:  well of the Plate
samtools_filter:  parameter used by samtools view to filter out badly mapped reads (e.g. -F2308 -f0x2 -q1 -> F2308_f0x2_q1)
pathname:  path to the idx_stats file

### From Excel spreadsheet that ArcDyn uses to record metadata for each sample
Full_name_of_the_sample:  full name of the sample from the ArcDyn metadata (e.g. 1999JUL08_Art3_TrapA_Wk27).
Sample_alias:  code for the Year_Trap_Week (e.g. 98_B_33)
ArcDyn_Plate_name:  ArcDyn plate name (e.g. Anteater, which is Plate A)
EI_Plate_name:  name of plate given by Earlham Institute (e.g. IPO3916)
TGAC_Barcode:  number on the barcode sticker that i placed on the plate (SAM27529_PRO 1323_S1_gDNA).  this links back to the photo i take of each plate after it is loaded.
Concentration_ng_per_ul:  DNA concentration as it is given to EI
Sample_volume_ul:  how much volume of liquid given to EI per sample in each well
Quantification method: the method ArcDyn used to quantify DNA amount
x.260_280_ratio:  the 260/280 ratio ArcDyn measured for the sample's DNA

### From bedtools/genomecov -d output file
sum_coverage:  sum of the coverage per nucleotide = the total bases of all the reads that mapped to each mitogenome
mean_coverage:  sum/length of the mitogenome that was mapped to
stddev:  standard deviation of the coverage (e.g. is the mapping even or uneven?)
coefvar:  stddev/mean_coverage.  another measure of unevenness
length:  length of the mitogenome being mapped to.  should be the same as mt_length
pathname_genomecov:  path to the bedtools genomecov -d output file

### Calculated columns
Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI:  reads that mapped to this COI spike sequence
Coleoptera_Elateridae_40ng_COI:  reads that mapped to this COI spike sequence
Coleoptera_Mordellidae_20ng_COI:  reads that mapped to this COI spike sequence
ratio_mordellid20_bombyx10:  ratio of mordellid to bombyx reads, should be 2
ratio_elaterid40_bombyx:  ratio of elaterid to bombyx reads, should be 4
ratio_elaterid40_mordellid20:  ratio of elaterid to mordellid reads, should be 2
elaterid_mordellid_sum:  sum of elaterid and mordellid reads
COI_corr:  COI_spike correction factor:  elaterid_mordellid_sum/min(elaterid_mordellid_sum)
lysis_buffer_proportion:  lysis buffer correction factor:  lysis_buffer_purified_ml/lysis_buffer_orig_total_ml
mapped_reads_COI_corr:  mapped reads after correction with COI_corr
mapped_reads_COI_lysis_corr:  mapped reads after correction with COI_corr and lysis_buffer



Don't use this code.  I wrote this when i tried to do the COI spike corrections with a wide table.  Wrong idea.   
```{r COI_spike corrections on wide table, eval=FALSE, include=FALSE}

# used for debugging, in case i screw up the dataset while writing code
# idx_meta_genomcov_mapped_reads_lysis_corr_wide <- read_delim("idx_meta_genomcov_mapped_reads_lysis_corr_wide.txt", delim = "\t", escape_double = FALSE, trim_ws = TRUE)

# pull out the COI spike rows
bombyx_corr <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% dplyr::filter(str_detect(.$mitogenome, '1-2_Lepidoptera_Bombycidae_Bombyx_mori_10ng_COI'))

mordellid_corr <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% dplyr::filter(str_detect(.$mitogenome, '3-1_Coleoptera_Mordellidae_20ng_COI'))

elaterid_corr <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% dplyr::filter(str_detect(.$mitogenome, '2-1_Coleoptera_Elateridae_40ng_COI'))

# calculate the ratios of the COI spikes and bind the rows to the original dataset
bombyx_bombyx <- as.data.frame(as.numeric(t(bombyx_corr[,-1]))/as.numeric(t(bombyx_corr[,-1])))
colnames(bombyx_bombyx)[1] <- "bombyx_bombyx"
bombyx_bombyx <- as.data.frame(t(bombyx_bombyx)) %>% rownames_to_column(var="mitogenome")

mordellid_bombyx <- data_frame(as.numeric(t(mordellid_corr[,-1]))/as.numeric(t(bombyx_corr[,-1])))
colnames(mordellid_bombyx)[1] <- "mordellid20_bombyx10"
mordellid_bombyx <- as.data.frame(t(mordellid_bombyx)) %>% rownames_to_column(var="mitogenome")

elaterid_bombyx <- data_frame(as.numeric(t(elaterid_corr[,-1]))/as.numeric(t(bombyx_corr[,-1])))
colnames(elaterid_bombyx)[1] <- "elaterid40_bombyx"
elaterid_bombyx <- as.data.frame(t(elaterid_bombyx)) %>% rownames_to_column(var="mitogenome")

elaterid_mordellid <- data_frame(as.numeric(t(elaterid_corr[,-1]))/as.numeric(t(mordellid_corr[,-1])))
colnames(elaterid_mordellid)[1] <- "elaterid40_mordellid20"
elaterid_mordellid <- as.data.frame(t(elaterid_mordellid)) %>% rownames_to_column(var="mitogenome")

# bind together and then bind onto bottom of genomecov wide dataset
COI_spike <- bind_rows(bombyx_bombyx, mordellid_bombyx, elaterid_bombyx, elaterid_mordellid)
colnames(COI_spike) <- colnames(idx_meta_genomcov_mapped_reads_lysis_corr_wide)
idx_meta_genomcov_mapped_reads_lysis_corr_wide <- bind_rows(idx_meta_genomcov_mapped_reads_lysis_corr_wide, COI_spike)

# check pairwise ratios of COI spikes
# should be 2, but it's 3.7
ratio_mordellid20_bombyx10 <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% filter(str_detect(.$mitogenome, 'mordellid20_bombyx10')) %>% dplyr::select(-mitogenome) %>% t() %>% as.numeric() %>% mean(na.rm = TRUE) %>% round(1); ratio_mordellid20_bombyx10

# should be 4, but it's 7.3
ratio_elaterid40_bombyx <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% filter(str_detect(.$mitogenome, 'elaterid40_bombyx')) %>% dplyr::select(-mitogenome) %>% t() %>% as.numeric() %>% mean(na.rm = TRUE) %>% round(1); ratio_elaterid40_bombyx

# should be 2, and it's 1.9
ratio_elaterid40_mordellid20 <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% filter(str_detect(.$mitogenome, 'elaterid40_mordellid20')) %>% dplyr::select(-mitogenome) %>% t() %>% as.numeric() %>% mean(na.rm = TRUE) %>% round(1); ratio_elaterid40_mordellid20

# Use the sum of elaterid and mordellid reads to do the correction.  Bombyx didn't show up

elaterid_mordellid_sum <- as.data.frame(as.numeric(t(elaterid_corr[,-1])) + as.numeric(t(mordellid_corr[,-1])))
colnames(elaterid_mordellid_sum)[1] <- "elaterid_mordellid_sum"
elaterid_mordellid_sum <- as.data.frame(t(elaterid_mordellid_sum)) %>% rownames_to_column(var="mitogenome")
colnames(elaterid_mordellid_sum) <- colnames(idx_meta_genomcov_mapped_reads_lysis_corr_wide)
idx_meta_genomcov_mapped_reads_lysis_corr_wide <- bind_rows(idx_meta_genomcov_mapped_reads_lysis_corr_wide, elaterid_mordellid_sum)

elaterid_mordellid_min <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% filter(str_detect(.$mitogenome, 'elaterid_mordellid_sum')) %>% dplyr::select(-mitogenome)  %>% t() %>% as.numeric() %>% min(na.rm = TRUE)
elaterid_mordellid_min # 115100.3 mapped_reads

# now calculate the column-wise correction factor COI_corr:  elaterid_mordellid_sum/elaterid_mordellid_min
COI_corr <- as.data.frame(as.numeric(t(elaterid_mordellid_sum[,-1]))/elaterid_mordellid_min)
colnames(COI_corr)[1] <- "COI_corr"
COI_corr <- as.data.frame(t(COI_corr)) %>% rownames_to_column(var="mitogenome")
colnames(COI_corr) <- colnames(idx_meta_genomcov_mapped_reads_lysis_corr_wide)
idx_meta_genomcov_mapped_reads_lysis_corr_wide <- bind_rows(idx_meta_genomcov_mapped_reads_lysis_corr_wide, COI_corr)


# START HERE:  pull out COI_corr and divide it into mapped_reads (using a long format dataset) and generate the new wide dataset with lysis and COI corrected reads
COI_corr_t <- t(COI_corr) %>% as.numeric()
%>% as.data.frame()
test <- idx_meta_genomcov_mapped_reads_lysis_corr_wide %>% gather(date, mapped_reads_lysis_corr, -mitogenome)


```
